{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "from dataclasses import dataclass, field\n",
    "from glob import glob\n",
    "import numpy as np\n",
    "import math\n",
    "import datetime as dt\n",
    "from functools import partial, reduce\n",
    "from typing import Callable\n",
    "import pandas as pd\n",
    "import pysam\n",
    "from operator import methodcaller\n",
    "from abc import ABC, abstractmethod\n",
    "from collections import Counter\n",
    "import re\n",
    "import scipy.stats as stats\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Processing(ABC):\n",
    "    '''Operations on pileup df'''\n",
    "    @abstractmethod\n",
    "    def operation():\n",
    "        pass\n",
    "\n",
    "class RemoveRedundantColumns(Processing):\n",
    "    '''Compute Max'''\n",
    "    \n",
    "    def operation(df: pd.DataFrame) -> pd.DataFrame:\n",
    "        for col in ['ref', 'seq']:\n",
    "            if not df[f'{col}_n'].equals(df[f'{col}_t']):\n",
    "                raise ValueError(f'{col} differs!')\n",
    "            df[col] = df[f'{col}_n']\n",
    "            del df[f'{col}_n']\n",
    "            del df[f'{col}_t']\n",
    "   \n",
    "        return df\n",
    "\n",
    "class RemoveRedundantColumns(Processing):\n",
    "    '''Compute Max'''\n",
    "    \n",
    "    def operation(df: pd.DataFrame) -> pd.DataFrame:\n",
    "        for col in ['ref', 'seq']:\n",
    "            if not df[f'{col}_n'].equals(df[f'{col}_t']):\n",
    "                raise ValueError(f'{col} differs!')\n",
    "            df[col] = df[f'{col}_n']\n",
    "            del df[f'{col}_n']\n",
    "            del df[f'{col}_t']\n",
    "   \n",
    "        return df\n",
    "\n",
    "\n",
    "def process_pile(df):\n",
    "    # __subclasses__ will found all classes inheriting from Operations\n",
    "    for step in Processing.__subclasses__():\n",
    "        df = step.operation(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'tmp_d' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb Cell 3\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m# def separate_indels_from_res(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39m#     def find_indels_in_res(res):\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m#         def extract_indel(res, val):\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=33'>34</a>\u001b[0m     \n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=34'>35</a>\u001b[0m \u001b[39m#     return df\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=36'>37</a>\u001b[0m ref_up \u001b[39m=\u001b[39m tmp_d\u001b[39m.\u001b[39mget(\u001b[39m'\u001b[39m\u001b[39mref\u001b[39m\u001b[39m'\u001b[39m)\u001b[39m.\u001b[39mupper()\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=37'>38</a>\u001b[0m to_keep \u001b[39m=\u001b[39m [\u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m,\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mT\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mC\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mG\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m*\u001b[39m\u001b[39m'\u001b[39m]  \u001b[39m# keep * cuz have phred score...\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W2sZmlsZQ%3D%3D?line=38'>39</a>\u001b[0m \u001b[39mif\u001b[39;00m i \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n",
      "\u001b[0;31mNameError\u001b[0m: name 'tmp_d' is not defined"
     ]
    }
   ],
   "source": [
    "# def separate_indels_from_res(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\n",
    "#     def find_indels_in_res(res):\n",
    "#         def extract_indel(res, val):\n",
    "#             indels = []\n",
    "#             bits = res.strip().split(val)\n",
    "#             if len(bits) > 1:\n",
    "#                 res = ''\n",
    "#                 for bit in bits:\n",
    "#                     if bit[0].isdigit():\n",
    "#                         indel_len = []\n",
    "#                         for char in bit:\n",
    "#                             if char.isdigit():\n",
    "#                                 indel_len.append(char)\n",
    "#                             else:\n",
    "#                                 break\n",
    "#                         indel_len = ''.join(indel_len)\n",
    "#                         indel = bit[len(indel_len):int(indel_len) + len(indel_len)]\n",
    "#                         indels.append(val + ':' + indel)\n",
    "#                         res += bit[len(indel_len) + int(indel_len):]\n",
    "#                     else:\n",
    "#                         res += bit\n",
    "#             return indels, res\n",
    "#         ins, res = extract_indel(res, '+')\n",
    "#         dels, res = extract_indel(res, '-')\n",
    "#         indels = ins+dels\n",
    "#         res = ''.join(char for char in res if char in '.,ATCGatcg*')\n",
    "#         most_seen_indel = Counter(indels).most_common()[0] if indels else 'NA'\n",
    "#         return pd.Series([res, most_seen_indel])\n",
    "#     df[[f'{sample_type}_indel_free_res', f'{sample_type}_indels']] = df[f'res_{sample_type}_no_carrot'].apply(find_indels_in_res)\n",
    "#     df[f'{sample_type}l1'] = df[f'{sample_type}_indel_free_res'].str.len() \n",
    "#     df[f'{sample_type}l2'] = df[f'qual_{sample_type}'].str.len()\n",
    "#     if not df[f'{sample_type}l1'].equals(df[f'{sample_type}l2']):\n",
    "#       raise ValueError(f'{sample_type}_indel_free_res and qual_{sample_type} should have same length!! {df[f\"{sample_type}l1\"].compare(df[f\"{sample_type}l1\"])}')\n",
    "    \n",
    "# #     return df\n",
    "\n",
    "# ref_up = tmp_d.get('ref').upper()\n",
    "# to_keep = ['.', ',', 'A', 'T', 'C', 'G', '*']  # keep * cuz have phred score...\n",
    "# if i == 0:\n",
    "#     res = ''.join([nuc for nuc in res if nuc.upper() in to_keep])\n",
    "#     res = res.replace('.', ref_up).replace(',', tmp_d.get('ref').lower())\n",
    "# else:\n",
    "#     res = ''.join([nuc for nuc in res if nuc.upper() in to_keep])\n",
    "#     res = res.replace('.', ref_up).replace(',', ref_up)\n",
    "# if not len(res) == len(tmp_d.get('qual')):\n",
    "#     print('not_same_len', len(res), len(tmp_d.get('qual')),\n",
    "#             res, tmp_d, i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [],
   "source": [
    "import operator\n",
    "\n",
    "\n",
    "Preprocessor = Callable[[pd.DataFrame, str], pd.DataFrame]\n",
    "\n",
    "class DataSchema:\n",
    "    AMOUNT = \"amount\"\n",
    "    CATEGORY = \"category\"\n",
    "    DATE = \"date\"\n",
    "    MONTH = \"month\"\n",
    "    YEAR = \"year\"\n",
    "\n",
    "class PileSchema:\n",
    "    AMOUNT = \"amount\"\n",
    "    CATEGORY = \"category\"\n",
    "    DATE = \"date\"\n",
    "    MONTH = \"month\"\n",
    "    YEAR = \"year\"\n",
    "\n",
    "def create_df(pile):#: str, chrom: str, start: int, end: int) -> pd.DataFrame:\n",
    "    '''Converts a genomic region of a pileup to a df'''\n",
    "    keys = ['seq', 'pos', 'ref', 'reads', 'res', 'qual', 'pos_in_read', 'read_names', 'flags', 'map']#TODO drop 'pos_in_read', 'read_names',\n",
    "    tabixfile = pysam.TabixFile(pile)\n",
    "    df = pd.DataFrame([line.split('\\t') for line in tabixfile.fetch()], columns=keys)\n",
    "    df.set_index('pos', inplace=True)\n",
    "    return df\n",
    "\n",
    "def remove_redundant_column(df: pd.DataFrame, col_name: str) -> pd.DataFrame:\n",
    "\n",
    "    if not df[f'{col_name}_n'].equals(df[f'{col_name}_t']):\n",
    "        raise ValueError(f'{col_name} differs!')\n",
    "    df[col_name] = df[f'{col_name}_n']\n",
    "    del df[f'{col_name}_n']\n",
    "    del df[f'{col_name}_t']\n",
    "   \n",
    "    return df\n",
    "\n",
    "def create_SV_column(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\n",
    "    df[f'SV_{sample_type}'] = df[f'res_{sample_type}'].str.count('^') + df[f'res_{sample_type}'].str.count('$')\n",
    "    return df\n",
    "\n",
    "def create_indel_column(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\n",
    "    df[f'SV_{sample_type}'] = df[f'res_{sample_type}'].str.count('^') + df[f'res_{sample_type}'].str.count('$')\n",
    "    return df\n",
    "\n",
    "def remove_carrots_from_res(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\n",
    "    # ^ (caret) marks the start of a read segment and the ASCII\n",
    "    # of the character following `^' minus 33 gives the mapping quality\n",
    "    def slice_and_dice(res):\n",
    "        if '^' in res:\n",
    "            if res.startswith('^'):\n",
    "                res = ''.join([bit[1:] for bit in res.split('^')])\n",
    "            else:\n",
    "                res = ''.join([bit[1:] if j != 0 else bit for j, bit in enumerate(res.split('^'))])\n",
    "        return res\n",
    "    df[f'res_{sample_type}_no_carrot'] = df[f'res_{sample_type}'].apply(slice_and_dice)\n",
    "    return df\n",
    "\n",
    "#def remove_low_quality(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\n",
    "def convert_to_upper(res, ref):\n",
    "    assert ref.isupper()\n",
    "    return res.replace(',', ref).replace('.', ref).replace('*', '').upper()\n",
    "\n",
    "def remove_ref_from_tumour_res(res_n, res_t, ref):\n",
    "\n",
    "    most_common_base_in_normal = Counter(convert_to_upper(res_n, ref)).most_common()[0][0]\n",
    "    return convert_to_upper(res_t, ref).replace(most_common_base_in_normal, '')\n",
    "\n",
    "def remove_positions_with_little_support_for_somatic_var(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    def putative_vars(res_t, res_n, ref):\n",
    "        # res_t = convert_to_upper(res_t, ref)\n",
    "        # res_n = convert_to_upper(res_n, ref)\n",
    "        # most_common_base_in_normal = Counter(res_n).most_common()[0][0]\n",
    "        res_t = remove_ref_from_tumour_res(res_n, res_t, ref)\n",
    "        if res_t:\n",
    "            return True if Counter(res_t).most_common()[0][1] > 4 else False\n",
    "        return False\n",
    "    return df[df.apply(lambda x: putative_vars(\n",
    "        x['res_t_no_carrot'],\n",
    "        x['res_n_no_carrot'],\n",
    "        x['ref']),\n",
    "        axis = 1\n",
    "    )]\n",
    "    \n",
    "#tried complied regex, still slow\n",
    "def separate_indels_from_res(df: pd.DataFrame, sample_type: str) -> pd.DataFrame:\n",
    "    def find_indels_in_res(res):\n",
    "        regex_result = re.findall(\"\\+[0-9]+[ACGTNacgtn]+|\\-[0-9]+[ACGTNacgtn]+\", res)\n",
    "        corected_regex_result = []#can't find other way to stop regex adding unrelated nucs at end\n",
    "        for indel in regex_result:\n",
    "            indel_len = ''.join(char for char in indel if char.isdigit())\n",
    "            corected_regex_result.append(indel[0] + indel_len + ''.join(char for char in indel if char.isalpha())[:int(indel_len)])\n",
    "        indel_count = Counter(corected_regex_result)\n",
    "        for indel in indel_count.keys():\n",
    "            res = ''.join(res.split(indel))\n",
    "        most_seen_indel = indel_count.most_common()[0] if indel_count else 'NA'\n",
    "        most_seen_indel_upper = np.nan if most_seen_indel == 'NA' else (most_seen_indel[0].upper(), most_seen_indel[1])\n",
    "        res = ''.join(char for char in res if char in '.,ATCGatcg*')\n",
    "        return pd.Series([res, most_seen_indel_upper])\n",
    "    df[[f'{sample_type}_indel_free_res', f'{sample_type}_indels']] = df[f'res_{sample_type}_no_carrot'].apply(find_indels_in_res)\n",
    "    if not df[f'{sample_type}_indel_free_res'].str.len().equals(df[f'qual_{sample_type}'].str.len()):\n",
    "      raise ValueError(f'{sample_type}_indel_free_res and qual_{sample_type} should have same length!! {df[f\"{sample_type}l2\"].compare(df[f\"{sample_type}l1\"])}')\n",
    "    return df\n",
    "\n",
    "def caller(count, normal_count, reads_t, reads_n, putative_somatic_var, vars):\n",
    "\n",
    "    vals = [count, normal_count]\n",
    "    nons = [int(reads_t) - count, int(reads_n) - normal_count]\n",
    "    oddsratio, pvalue1 = stats.fisher_exact([vals, nons], alternative='greater')\n",
    "    if pvalue1 < 0.01:\n",
    "        vars.append((putative_somatic_var, pvalue1))\n",
    "    return vars\n",
    "    \n",
    "def call_var(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    def test_putative_somatic_SNVs(tumour_nucs, normal_nucs, reads_t, reads_n, ref, res_n):\n",
    "        normal_nucs_upper = convert_to_upper(normal_nucs, ref)\n",
    "        tumour_nucs_no_ref_upper = remove_ref_from_tumour_res(normal_nucs, tumour_nucs, ref)\n",
    "        normal_nucs_most_common = dict(Counter(normal_nucs_upper).most_common())\n",
    "        tumour_nucs_most_common = {\n",
    "            t_var: count for t_var, count in Counter(tumour_nucs_no_ref_upper).items() if normal_nucs_most_common.get(t_var, 0) < 2\n",
    "        }\n",
    "        vars = []\n",
    "        for putative_somatic_var, count in tumour_nucs_most_common.items():\n",
    "            if count > 4:\n",
    "                normal_count = normal_nucs_most_common.get(putative_somatic_var, 0)\n",
    "                vars = caller(count, normal_count, reads_t, reads_n, putative_somatic_var, vars)\n",
    "        return vars if vars else np.nan\n",
    "    df[\"SNV_calls\"] = df.apply(lambda x: test_putative_somatic_SNVs(\n",
    "        x['t_indel_free_res'],\n",
    "        x['n_indel_free_res'],\n",
    "        x['reads_t'],\n",
    "        x['reads_n'],\n",
    "        x['ref'],\n",
    "        x['res_n']),\n",
    "        axis = 1\n",
    "    )\n",
    "    def test_top_putative_somatic_indel(tumour_indels, reads_t, reads_n, ref, res_n):\n",
    "        vars = []\n",
    "        if type(tumour_indels) != float:\n",
    "            freq_of_most_common_tumour_indel_in_normal = res_n.upper().count(tumour_indels[0])\n",
    "            if freq_of_most_common_tumour_indel_in_normal < 2:\n",
    "                normal_d = dict([(tumour_indels[0], freq_of_most_common_tumour_indel_in_normal)])\n",
    "                putative_somatic_var, count = tumour_indels\n",
    "                if count > 4:\n",
    "                    normal_count = normal_d.get(putative_somatic_var, 0)\n",
    "                    vars = caller(count, normal_count, reads_t, reads_n, putative_somatic_var, vars)\n",
    "        return vars if vars else np.nan\n",
    "    df[\"indel_call\"] = df.apply(lambda x: test_top_putative_somatic_indel(\n",
    "        x['t_indels'],\n",
    "        x['reads_t'],\n",
    "        x['reads_n'],\n",
    "        x['ref'],\n",
    "        x['res_n']),\n",
    "        axis = 1\n",
    "    )\n",
    "    def test_top_putative_somatic_SV(tumour_SVs, normal_SVs, reads_t, reads_n, ref, res_n):\n",
    "        #no limit as some reads naturally start and stop....\n",
    "        vars = []\n",
    "        if tumour_SVs > 4:\n",
    "            vars = caller(tumour_SVs, normal_SVs, reads_t, reads_n, 'SV', vars)\n",
    "        return vars if vars else np.nan\n",
    "    df[\"SV_call\"] = df.apply(lambda x: test_top_putative_somatic_SV(\n",
    "        x['SV_t'],\n",
    "        x['SV_n'],\n",
    "        x['reads_t'],\n",
    "        x['reads_n'],\n",
    "        x['ref'],\n",
    "        x['res_n']),\n",
    "        axis = 1\n",
    "    )\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(98038, 18)\n",
      "(20488, 20)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(57, 27)"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tumour = create_df('/Users/liam/code/personal/SBSC/tests/resources/HCC1937_t_RNF157.pileup.gz')\n",
    "df_normal = create_df('/Users/liam/code/personal/SBSC/tests/resources/HCC1937_n_RNF157.pileup.gz')\n",
    "df = df_normal.join(df_tumour, how='inner', lsuffix='_n', rsuffix='_t')\n",
    "print(df.shape)\n",
    "df = remove_redundant_column(df, 'ref')\n",
    "df = remove_redundant_column(df, 'seq')\n",
    "df = create_SV_column(df, 't')\n",
    "df = create_SV_column(df, 'n')\n",
    "df = remove_carrots_from_res(df, 't')\n",
    "df = remove_carrots_from_res(df, 'n')\n",
    "df = remove_positions_with_little_support_for_somatic_var(df)\n",
    "print(df.shape)\n",
    "df = separate_indels_from_res(df, 't')\n",
    "df = separate_indels_from_res(df, 'n')\n",
    "df = call_var(df)\n",
    "df = df[df[['SNV_calls', 'indel_call', 'SV_call']].notna().any(1)]\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "TumourSchema.SV() takes 0 positional arguments but 1 was given",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb Cell 6\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W5sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mTumourSchema\u001b[39m.\u001b[39mSTRUCTURAL_VARIANT\u001b[39m}\u001b[39;00m\u001b[39m_tumour\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W5sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m t\u001b[39m=\u001b[39mTumourSchema()\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/liam/code/personal/SBSC/SBSC/Untitled-1.ipynb#W5sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m t\u001b[39m.\u001b[39;49mSV\n",
      "\u001b[0;31mTypeError\u001b[0m: TumourSchema.SV() takes 0 positional arguments but 1 was given"
     ]
    }
   ],
   "source": [
    "@dataclass \n",
    "class PileupSchema:#TODO ???drop 'pos_in_read', 'read_names'????\n",
    "    SEQUENCE = \"seq\"\n",
    "    POSITION = \"pos\"\n",
    "    REFERENCE = \"ref\"\n",
    "    READS = \"reads\"\n",
    "    RESULTS = \"res\"\n",
    "    QUALITY = \"qual\"\n",
    "    POSITION_IN_READ = \"pos_in_read\"\n",
    "    FLAGS = \"flags\"\n",
    "    MAPPING_QUALITY = \"map\"\n",
    "\n",
    "@dataclass\n",
    "class ProcessedSchema(PileupSchema):\n",
    "    STRUCTURAL_VARIANT = \"SV\"\n",
    "\n",
    "@dataclass\n",
    "class TumourSchema(ProcessedSchema):\n",
    "    \n",
    "    @property\n",
    "    def SV():\n",
    "        return f'{TumourSchema.STRUCTURAL_VARIANT}_tumour'\n",
    "t=TumourSchema()\n",
    "t.SV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: ref, dtype: object)"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = df.ref.index.get_loc('76145432')\n",
    "\n",
    "df.ref.iloc[idx - 10 : idx + 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DataSchema:\n",
    "    AMOUNT = \"amount\"\n",
    "    CATEGORY = \"category\"\n",
    "    DATE = \"date\"\n",
    "    MONTH = \"month\"\n",
    "    YEAR = \"year\"\n",
    "\n",
    "   \n",
    "\n",
    "def create_df(pile):#: str, chrom: str, start: int, end: int) -> pd.DataFrame:\n",
    "    '''Converts a genomic region of a pileup to a df'''\n",
    "    keys = ['seq', 'pos', 'ref', 'reads', 'res', 'qual', 'pos_in_read', 'read_names', 'flags', 'map']#TODO drop 'pos_in_read', 'read_names',\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['amount', 'category', 'date', 'month', 'year']"
      ]
     },
     "execution_count": 351,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a= [name for schema, name in vars(DataSchema).items() if not schema.startswith(\"__\")]\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('tumour_amount', 'normal_pos_in_read')"
      ]
     },
     "execution_count": 377,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "class PileupSchema:#TODO ???drop 'pos_in_read', 'read_names'????\n",
    "    SEQUENCE = \"seq\"\n",
    "    POSITION = \"pos\"\n",
    "    REFERENCE = \"ref\"\n",
    "    READS = \"reads\"\n",
    "    RESULTS = \"res\"\n",
    "    QUALITY = \"qual\"\n",
    "    POSITION_IN_READ = \"pos_in_read\"\n",
    "    FLAGS = \"flags\"\n",
    "    MAPPING_QUALITY = \"map\"\n",
    "\n",
    "\n",
    "class ProcessedDataSchema(PileupSchema):\n",
    "    AMOUNT = \"amount\"\n",
    "    CATEGORY = \"category\"\n",
    "    DATE = \"date\"\n",
    "    MONTH = \"month\"\n",
    "    YEAR = \"year\"\n",
    "\n",
    "    \n",
    "def ff(sample_type, name: ProcessedDataSchema):\n",
    "    return f'{sample_type}_{name}'\n",
    "\n",
    "tumour_cols = partial(ff, 'tumour')\n",
    "normal_cols = partial(ff, 'normal')\n",
    "\n",
    "tumour_cols(ProcessedDataSchema.AMOUNT), normal_cols(ProcessedDataSchema.POSITION_IN_READ)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['seq', 'pos', 'ref', 'reads', 'res', 'qual', 'pos_in_read', 'flags', 'map']\n"
     ]
    }
   ],
   "source": [
    "p=PileupSchema()\n",
    "print(PileupSchema.list_all_columns())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df[['SNV_calls', 'indel_call', 'SV_call']].notna().any(1)][['res_t_no_carrot','res_n_no_carrot', 'SNV_calls', 'indel_call', 'SV_call', 't_indels']].to_csv('~/Downloads/regex.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seq_n                                                        chr17\n",
       "ref_n                                                            C\n",
       "reads_n                                                         59\n",
       "res_n            *A,,*.,,,.,,.*,*,,.,,.*,,.**,,,,.,*tAA,,,t,.*A...\n",
       "qual_n           ('8/)9=9&>848/4287@5B:&&36%-).;82-/47(941';8%4...\n",
       "pos_in_read_n    14876,13621,13831,13241,9962,9469,8150,7546,76...\n",
       "read_names_n     a7f99129-4d5a-42b1-994b-4e4fcce73ce1,5e17a39a-...\n",
       "flags_n          0,0,16,16,16,0,16,16,16,0,16,16,0,0,16,0,16,16...\n",
       "map_n            WY[\\XY[ZY]YZWNX]]]Z\\[ZXY[ZRW[WS]YT]SZ]]X]]]Z[]...\n",
       "seq_t                                                        chr17\n",
       "ref_t                                                            C\n",
       "reads_t                                                         51\n",
       "res_t            *AGt,,,a.A.,,AA,,A.A,*,..,,*A,A,A,,A,.A*,,,aAa...\n",
       "qual_t           '5%$<4766(+8<1.;57007&1%8273-//3:-46,74868+>-?...\n",
       "pos_in_read_t    15568,10657,9258,7684,7560,7148,7008,6871,7001...\n",
       "read_names_t     9308d7ef-3936-425a-b542-42725e0bfbac,14cd153f-...\n",
       "flags_t          0,0,0,16,16,16,16,16,0,0,0,16,16,0,0,16,16,0,0...\n",
       "map_t            ]ZQYWYXXZYYVTY[Z]Y[\\][[XWV]ZWZ]YXZ]YZ]OJ]XWN]W...\n",
       "Name: 76157884, dtype: object"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc['76157884',:]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1\n",
      "+1\n",
      "+1\n",
      "-2t\n",
      "-2t\n",
      "-11tttttttttt\n",
      "-1\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#failed attempt for do regex faster\n",
    "chars, index = [], []\n",
    "for i, char in enumerate(a):\n",
    "    if char in '+-':\n",
    "        chars.append('|')\n",
    "        index.append(i)\n",
    "    if char.isdigit():\n",
    "        chars.append(char)\n",
    "        index.append(i)\n",
    "\n",
    "kk = ''.join(chars)\n",
    "index = [pos for i, pos in enumerate(index) if kk[i] == '|']\n",
    "assert len(kk.split('|')[1:]) == len(index)\n",
    "indels = kk.split('|')[1:]\n",
    "for i, indel in zip(index, indels):\n",
    "    print(a[(i): i+len(indel)+int(indel)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ",,,.*.,,-1t,.,.....,,..-2TT.,-1t,.,.-1T.-1T,,....\t..-2TT,,*....+2TT.-2TT,-1t,-1t....,,+1t.*...+1...\t,,,.*.,,,.,.....,,...,,.,..,,.,..,..,.,,,..,,....\tNA\t44\t44\t..,,*.....,,....,,.*......,.,.,.,,,,,.,,...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['-1t', '-2TT', '-1t', '-1T', '-1T']"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.findall(\"\\+[0-9]+[ACGTNacgtn]+|\\-[0-9]+[ACGTNacgtn]+\", ',,,.*.,,-1t,.,.....,,..-2TT.,-1t,.,.-1T.-1T,,....')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reads_n</th>\n",
       "      <th>res_n</th>\n",
       "      <th>qual_n</th>\n",
       "      <th>pos_in_read_n</th>\n",
       "      <th>read_names_n</th>\n",
       "      <th>flags_n</th>\n",
       "      <th>map_n</th>\n",
       "      <th>reads_t</th>\n",
       "      <th>res_t</th>\n",
       "      <th>qual_t</th>\n",
       "      <th>...</th>\n",
       "      <th>SV_t</th>\n",
       "      <th>SV_n</th>\n",
       "      <th>res_n_no_carrot</th>\n",
       "      <th>res_t_no_carrot</th>\n",
       "      <th>t_indel_free_res</th>\n",
       "      <th>t_ins</th>\n",
       "      <th>t_dels</th>\n",
       "      <th>n_indel_free_res</th>\n",
       "      <th>n_ins</th>\n",
       "      <th>n_dels</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76142500</th>\n",
       "      <td>45</td>\n",
       "      <td>..*,......,,....,,........,.,.,.,,,,,.,,....^S,</td>\n",
       "      <td>9F#I%F3D6B&gt;I1C&gt;B-IDE7=AFD&lt;&gt;90;&gt;A3882FF@7D=EA;</td>\n",
       "      <td>9114,8900,7709,7439,6783,6281,5580,5562,7117,5...</td>\n",
       "      <td>c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...</td>\n",
       "      <td>0,0,16,16,0,0,0,0,0,0,16,16,0,0,0,0,16,16,0,0,...</td>\n",
       "      <td>XXY[TZRYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[&lt;]V]XVWY[XS</td>\n",
       "      <td>52</td>\n",
       "      <td>,,,...,,.$,....,,...,,.,..,,.,..,..,.,,,..,,.....</td>\n",
       "      <td>38&gt;A479:B7B&lt;44G2@CA+B+475&gt;298@=75D&gt;1&lt;?;3(9@5=?...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>..*,......,,....,,........,.,.,.,,,,,.,,....,</td>\n",
       "      <td>,,,...,,.$,....,,...,,.,..,,.,..,..,.,,,..,,.....</td>\n",
       "      <td>,,,...,,.$,....,,...,,.,..,,.,..,..,.,,,..,,.....</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>..*,......,,....,,........,.,.,.,,,,,.,,....,</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76142565</th>\n",
       "      <td>45</td>\n",
       "      <td>..,,.....,,....,,T.......t.,.,.,,,,,.,,....,^Z,</td>\n",
       "      <td>&lt;*;6,'&amp;&amp;137+916?A'0'9)C*/%73'8875;4/&gt;.5'&lt;+*16</td>\n",
       "      <td>9178,8968,7762,7503,6844,5646,5627,7183,5344,5...</td>\n",
       "      <td>c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...</td>\n",
       "      <td>0,0,16,16,0,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,...</td>\n",
       "      <td>XXY[TRYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[&lt;]V]XVWY[XSZ</td>\n",
       "      <td>51</td>\n",
       "      <td>,,,...,,,....,,..G,,.,-1g..,,.,..,..G,,,..,,.....</td>\n",
       "      <td>764J0159&lt;A)./=7(%(;.'$#$%6&amp;)72:.5$'7=(/&lt;7,30./...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>..,,.....,,....,,T.......t.,.,.,,,,,.,,....,,</td>\n",
       "      <td>,,,...,,,....,,..G,,.,-1g..,,.,..,..G,,,..,,.....</td>\n",
       "      <td>,,,...,,,....,,..G,,.,..,,.,..,..G,,,..,,........</td>\n",
       "      <td>[]</td>\n",
       "      <td>[g]</td>\n",
       "      <td>..,,.....,,....,,T.......t.,.,.,,,,,.,,....,,</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76142572</th>\n",
       "      <td>46</td>\n",
       "      <td>.-1C.,,.-1C...-1C.,,....-1C,,.......-1C.,.,.,+...</td>\n",
       "      <td>@?/%;957A,;B=1&gt;46&gt;=@&lt;'C??$F60,@1)9&amp;&amp;&lt;%)&gt;;=1&amp;74</td>\n",
       "      <td>9185,8975,7769,7510,6851,5653,5634,7190,5351,5...</td>\n",
       "      <td>c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...</td>\n",
       "      <td>0,0,16,16,0,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,...</td>\n",
       "      <td>XXY[TRYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[&lt;]V]XVWY[XSZ[</td>\n",
       "      <td>51</td>\n",
       "      <td>,,,..-1C.c,,...-1C.,,...,,.,..,,.,..,...,*,.+1...</td>\n",
       "      <td>?&amp;+=&gt;:+$'@:594=&gt;=.(38$93-$&gt;(A7-:=&lt;$(?9@$.&gt;@&lt;:D...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>.-1C.,,.-1C...-1C.,,....-1C,,.......-1C.,.,.,+...</td>\n",
       "      <td>,,,..-1C.c,,...-1C.,,...,,.,..,,.,..,...,*,.+1...</td>\n",
       "      <td>,,,...c,,....,,...,,.,..,,.,..,...,*,..,,........</td>\n",
       "      <td>[C]</td>\n",
       "      <td>[C, C, C, C, C]</td>\n",
       "      <td>..,,.....,,....,,........,.,.,.,*,**.,*....,,,</td>\n",
       "      <td>[c]</td>\n",
       "      <td>[C, C, C, C, C, C, c]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76142791</th>\n",
       "      <td>46</td>\n",
       "      <td>..,,....,,....,,........,.,.,.,,,,,.,,....,c,^W.</td>\n",
       "      <td>=&lt;89/7&lt;;7:433;7@:7:&lt;5=6@@84(;8?::;;7&lt;399=;;#:8</td>\n",
       "      <td>9399,9189,7989,7722,7057,5850,7402,5563,5476,5...</td>\n",
       "      <td>c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...</td>\n",
       "      <td>0,0,16,16,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,0,...</td>\n",
       "      <td>XXY[TYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[&lt;]V]XVWY[XSZ[W</td>\n",
       "      <td>50</td>\n",
       "      <td>,,...,,-2at,...*,,...,,.,..-3ATG,,.,..,...,,,....</td>\n",
       "      <td>:6,&amp;=?)&gt;&gt;;&lt;(95&lt;0&lt;89:=?&amp;29;&lt;7;&lt;@:5+;996?86&lt;=49?...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>..,,....,,....,,........,.,.,.,,,,,.,,....,c,.</td>\n",
       "      <td>,,...,,-2at,...*,,...,,.,..-3ATG,,.,..,...,,,....</td>\n",
       "      <td>,,...,,,...*,,...,,.,..,,.,..,...,,,..,,.....,...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[at, ATG]</td>\n",
       "      <td>..,,....,,....,,........,.,.,.,,,,,.,,....,c,.</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76142869</th>\n",
       "      <td>47</td>\n",
       "      <td>.*,,-1g..G.,,....,,........,.,.,.,,,,,-1g.,,.....</td>\n",
       "      <td>'&amp;;92;*.?*,1,01@)944),8;-3($85(67:&amp;&gt;57$41;+;932</td>\n",
       "      <td>9475,9265,8066,7803,7136,5928,7478,5635,5555,5...</td>\n",
       "      <td>c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...</td>\n",
       "      <td>0,0,16,16,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,0,...</td>\n",
       "      <td>XXY[TYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[&lt;]V]XVWY[XSZ[WU</td>\n",
       "      <td>51</td>\n",
       "      <td>,,...,,,G...,,...,*T,G.,,.,..t...g,,..,,....-1...</td>\n",
       "      <td>*99+.)?9/&amp;),&lt;)20&amp;20(+$).'1&lt;=-170/%:7+(%/-*.%&lt;-...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>.*,,-1g..G.,,....,,........,.,.,.,,,,,-1g.,,.....</td>\n",
       "      <td>,,...,,,G...,,...,*T,G.,,.,..t...g,,..,,....-1...</td>\n",
       "      <td>,,...,,,G...,,...,*T,G.,,.,..t...g,,..,,.....g...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[G]</td>\n",
       "      <td>.*,,..G.,,....,,........,.,.,.,,,,,.,,....,,,..</td>\n",
       "      <td>[]</td>\n",
       "      <td>[g, g]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76239175</th>\n",
       "      <td>45</td>\n",
       "      <td>,...,.,,,,,,,,A.,.,.,,.+2AA,,A,,....,..,..,,,....</td>\n",
       "      <td>&gt;5=5=-5%/67@;6,9B3=:=9'A)%&gt;?+'.$=5&lt;=A-&lt;:?,&amp;2)</td>\n",
       "      <td>10933,10525,8979,8617,7688,7588,6637,6588,6117...</td>\n",
       "      <td>d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...</td>\n",
       "      <td>16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,0,2064...</td>\n",
       "      <td>VU[TW]UY[TYZU[]]WQZNZ[[[]]ZT[]QZ]]]Z[YXWZ]P]W</td>\n",
       "      <td>52</td>\n",
       "      <td>..,,,,.*,,..,.,,,,,..,,....,..,..,,,,+1t.,..C....</td>\n",
       "      <td>';9BB?6$=..-.3:-6?./A&lt;?&amp;()6/9.==+&lt;(&amp;%$=,/#&lt;;F1...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>,...,.,,,,,,,,A.,.,.,,.+2AA,,A,,....,..,..,,,...,</td>\n",
       "      <td>..,,,,.*,,..,.,,,,,..,,....,..,..,,,,+1t.,..C....</td>\n",
       "      <td>..,,,,.*,,..,.,,,,,..,,....,..,..,,,,.,..C.,.,...</td>\n",
       "      <td>[t]</td>\n",
       "      <td>[]</td>\n",
       "      <td>,...,.,,,,,,,,A.,.,.,,.,,A,,....,..,..,,,...,</td>\n",
       "      <td>[AA]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76239176</th>\n",
       "      <td>46</td>\n",
       "      <td>,...,.,t,,,,,,..-2GA,.,.,,.,,.,,...-1G.,..,..,...</td>\n",
       "      <td>D&lt;97;06$16;C&lt;@)4B5B;?;&amp;@5':A-&amp;.%?&lt;=AA-8=?/(*5%</td>\n",
       "      <td>10934,10526,8980,8618,7689,7589,6638,6589,6118...</td>\n",
       "      <td>d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...</td>\n",
       "      <td>16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,0,2064...</td>\n",
       "      <td>VU[TW]UY[TYZU[]]WQZNZ[[[]]ZT[]QZ]]]Z[YXWZ]P]W]</td>\n",
       "      <td>52</td>\n",
       "      <td>..,,,,.*,,.-1G.,.,,,,,..,,....,..,..,,,,.+1G,....</td>\n",
       "      <td>-4;@CB&lt;$&lt;8*)+4829A80C=C%()70;099.?+(/$=56%:@:/...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>,...,.,t,,,,,,..-2GA,.,.,,.,,.,,...-1G.,..,..,...</td>\n",
       "      <td>..,,,,.*,,.-1G.,.,,,,,..,,....,..,..,,,,.+1G,....</td>\n",
       "      <td>..,,,,.*,,..,.,,,,,..,,....,..,..,,,,.,....,.,...</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[G]</td>\n",
       "      <td>,...,.,t,,,,,,..,.,.,,.,,.,,....,..,..,,,...,,</td>\n",
       "      <td>[]</td>\n",
       "      <td>[GA, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76239270</th>\n",
       "      <td>46</td>\n",
       "      <td>,...,.+1T,,,*,,,,..,.,.+1T,,.,,.,,...,..,.+1T*...</td>\n",
       "      <td>-6&amp;-:&amp;9(&gt;&amp;41/'2+(*;$)8315':;+&amp;8.;')$&amp;',&amp;2%(;8*</td>\n",
       "      <td>11035,10619,9074,8712,7784,7678,6733,6684,6212...</td>\n",
       "      <td>d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...</td>\n",
       "      <td>16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,0,2064...</td>\n",
       "      <td>VU[TW]UY[TYZU[]]WQZNZ[[[]]ZT[]Z]]]Z[YXWZ]P]W]]</td>\n",
       "      <td>54</td>\n",
       "      <td>..,,,,.,,,..,.,,,,,.+1G.,,....,..-2AT,..,,,,.,...</td>\n",
       "      <td>%9'9.?%(5&amp;6*,77$,1&amp;%4&amp;'*$''($%&gt;%0&lt;)?&amp;*5*+*&lt;,39...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>,...,.+1T,,,*,,,,..,.,.+1T,,.,,.,,...,..,.+1T*...</td>\n",
       "      <td>..,,,,.,,,..,.,,,,,.+1G.,,....,..-2AT,..,,,,.,...</td>\n",
       "      <td>..,,,,.,,,..,.,,,,,..,,....,..,..,,,,.,.**.,.,...</td>\n",
       "      <td>[G]</td>\n",
       "      <td>[AT, A, A]</td>\n",
       "      <td>,...,.,,,*,,,,..,.,.,,.,,.,,...,..,.*,,t...,,,</td>\n",
       "      <td>[T, T, T]</td>\n",
       "      <td>[A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76239550</th>\n",
       "      <td>44</td>\n",
       "      <td>,*.*,**,,,,,,,.**,*,,.,,.,+5ggttt,*.A,**,..,,*...</td>\n",
       "      <td>:=&lt;%28&lt;8@2(4+::5-0&amp;9)9&lt;'..?35,??:6&gt;:&lt;K&gt;&lt;$,;D</td>\n",
       "      <td>11324,10887,9344,8984,8060,7956,7014,6955,6495...</td>\n",
       "      <td>d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...</td>\n",
       "      <td>16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,2064,0...</td>\n",
       "      <td>VU[TW]UY[TYZU[]WQZNZ[[[]]ZT[]Z]]]Z[YXWZ]P]W]</td>\n",
       "      <td>55</td>\n",
       "      <td>*A**,*.,,t*.,**,,*..*a..*****,*.,,,.*.*..**,*....</td>\n",
       "      <td>-+0'&gt;?@&lt;7(7C5:;/9+C=57&lt;B7%:61?&gt;=:&gt;&amp;99;)1:4&lt;;=:...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>,*.*,**,,,,,,,.**,*,,.,,.,+5ggttt,*.A,**,..,,*...</td>\n",
       "      <td>*A**,*.,,t*.,**,,*..*a..*****,*.,,,.*.*..**,*....</td>\n",
       "      <td>*A**,*.,,t*.,**,,*..*a..*****,*.,,,.*.*..**,*....</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>,*.*,**,,,,,,,.**,*,,.,,.,,*.A,**,..,,*.*.,.</td>\n",
       "      <td>[ggttt]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76240015</th>\n",
       "      <td>42</td>\n",
       "      <td>,..,.,,,,,,,.,.,.,t.,,.,+3aag,A..,..,..,t,..,.^V,</td>\n",
       "      <td>=31C&gt;7/)&amp;%,0+&lt;&gt;24-%@661,7$,9*-/4&gt;D3H*4&lt;756</td>\n",
       "      <td>11792,11342,9431,8515,8411,7477,7415,6958,6577...</td>\n",
       "      <td>d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...</td>\n",
       "      <td>16,0,0,16,0,16,16,16,16,16,16,16,0,2064,0,16,0...</td>\n",
       "      <td>VUTW]UY[TYZ[]WQZNZ[[[]]ZT[]Z]]]Z[YXWZ]PW]V</td>\n",
       "      <td>50</td>\n",
       "      <td>GA,,,,+2aat,,.*,G,,,..*...,...+1A.,,a.,.+1G..+...</td>\n",
       "      <td>%(-$'(&amp;%%$51*+&amp;-6&amp;,)-.'%5/(%4#4-43.39'%'%.',7-...</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>,..,.,,,,,,,.,.,.,t.,,.,+3aag,A..,..,..,t,..,.,</td>\n",
       "      <td>GA,,,,+2aat,,.*,G,,,..*...,...+1A.,,a.,.+1G..+...</td>\n",
       "      <td>GA,,,,t,,.*,G,,,..*...,....,,a.,.....,gAT..,.....</td>\n",
       "      <td>[aa, A, G, G, G]</td>\n",
       "      <td>[]</td>\n",
       "      <td>,..,.,,,,,,,.,.,.,t.,,.,,A..,..,..,t,..,.,</td>\n",
       "      <td>[aag]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>759 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reads_n                                              res_n  \\\n",
       "pos                                                                   \n",
       "76142500      45    ..*,......,,....,,........,.,.,.,,,,,.,,....^S,   \n",
       "76142565      45    ..,,.....,,....,,T.......t.,.,.,,,,,.,,....,^Z,   \n",
       "76142572      46  .-1C.,,.-1C...-1C.,,....-1C,,.......-1C.,.,.,+...   \n",
       "76142791      46   ..,,....,,....,,........,.,.,.,,,,,.,,....,c,^W.   \n",
       "76142869      47  .*,,-1g..G.,,....,,........,.,.,.,,,,,-1g.,,.....   \n",
       "...          ...                                                ...   \n",
       "76239175      45  ,...,.,,,,,,,,A.,.,.,,.+2AA,,A,,....,..,..,,,....   \n",
       "76239176      46  ,...,.,t,,,,,,..-2GA,.,.,,.,,.,,...-1G.,..,..,...   \n",
       "76239270      46  ,...,.+1T,,,*,,,,..,.,.+1T,,.,,.,,...,..,.+1T*...   \n",
       "76239550      44  ,*.*,**,,,,,,,.**,*,,.,,.,+5ggttt,*.A,**,..,,*...   \n",
       "76240015      42  ,..,.,,,,,,,.,.,.,t.,,.,+3aag,A..,..,..,t,..,.^V,   \n",
       "\n",
       "                                                   qual_n  \\\n",
       "pos                                                         \n",
       "76142500    9F#I%F3D6B>I1C>B-IDE7=AFD<>90;>A3882FF@7D=EA;   \n",
       "76142565    <*;6,'&&137+916?A'0'9)C*/%73'8875;4/>.5'<+*16   \n",
       "76142572   @?/%;957A,;B=1>46>=@<'C??$F60,@1)9&&<%)>;=1&74   \n",
       "76142791   =<89/7<;7:433;7@:7:<5=6@@84(;8?::;;7<399=;;#:8   \n",
       "76142869  '&;92;*.?*,1,01@)944),8;-3($85(67:&>57$41;+;932   \n",
       "...                                                   ...   \n",
       "76239175    >5=5=-5%/67@;6,9B3=:=9'A)%>?+'.$=5<=A-<:?,&2)   \n",
       "76239176   D<97;06$16;C<@)4B5B;?;&@5':A-&.%?<=AA-8=?/(*5%   \n",
       "76239270   -6&-:&9(>&41/'2+(*;$)8315':;+&8.;')$&',&2%(;8*   \n",
       "76239550     :=<%28<8@2(4+::5-0&9)9<'..?35,??:6>:<K><$,;D   \n",
       "76240015       =31C>7/)&%,0+<>24-%@661,7$,9*-/4>D3H*4<756   \n",
       "\n",
       "                                              pos_in_read_n  \\\n",
       "pos                                                           \n",
       "76142500  9114,8900,7709,7439,6783,6281,5580,5562,7117,5...   \n",
       "76142565  9178,8968,7762,7503,6844,5646,5627,7183,5344,5...   \n",
       "76142572  9185,8975,7769,7510,6851,5653,5634,7190,5351,5...   \n",
       "76142791  9399,9189,7989,7722,7057,5850,7402,5563,5476,5...   \n",
       "76142869  9475,9265,8066,7803,7136,5928,7478,5635,5555,5...   \n",
       "...                                                     ...   \n",
       "76239175  10933,10525,8979,8617,7688,7588,6637,6588,6117...   \n",
       "76239176  10934,10526,8980,8618,7689,7589,6638,6589,6118...   \n",
       "76239270  11035,10619,9074,8712,7784,7678,6733,6684,6212...   \n",
       "76239550  11324,10887,9344,8984,8060,7956,7014,6955,6495...   \n",
       "76240015  11792,11342,9431,8515,8411,7477,7415,6958,6577...   \n",
       "\n",
       "                                               read_names_n  \\\n",
       "pos                                                           \n",
       "76142500  c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...   \n",
       "76142565  c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...   \n",
       "76142572  c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...   \n",
       "76142791  c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...   \n",
       "76142869  c99bf8d8-0be9-4ba9-8c41-4c49ba953f39,40364381-...   \n",
       "...                                                     ...   \n",
       "76239175  d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...   \n",
       "76239176  d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...   \n",
       "76239270  d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...   \n",
       "76239550  d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...   \n",
       "76240015  d4e6733f-0b2e-4dae-b6bb-50906aae9d3f,c6084eeb-...   \n",
       "\n",
       "                                                    flags_n  \\\n",
       "pos                                                           \n",
       "76142500  0,0,16,16,0,0,0,0,0,0,16,16,0,0,0,0,16,16,0,0,...   \n",
       "76142565  0,0,16,16,0,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,...   \n",
       "76142572  0,0,16,16,0,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,...   \n",
       "76142791  0,0,16,16,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,0,...   \n",
       "76142869  0,0,16,16,0,0,0,0,16,16,0,0,0,0,16,16,0,0,0,0,...   \n",
       "...                                                     ...   \n",
       "76239175  16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,0,2064...   \n",
       "76239176  16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,0,2064...   \n",
       "76239270  16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,0,2064...   \n",
       "76239550  16,0,0,0,16,0,16,16,16,16,16,16,16,16,0,2064,0...   \n",
       "76240015  16,0,0,16,0,16,16,16,16,16,16,16,0,2064,0,16,0...   \n",
       "\n",
       "                                                    map_n reads_t  \\\n",
       "pos                                                                 \n",
       "76142500    XXY[TZRYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[<]V]XVWY[XS      52   \n",
       "76142565    XXY[TRYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[<]V]XVWY[XSZ      51   \n",
       "76142572   XXY[TRYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[<]V]XVWY[XSZ[      51   \n",
       "76142791   XXY[TYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[<]V]XVWY[XSZ[W      50   \n",
       "76142869  XXY[TYVWW[XYWZT]ZYYZRX[]Y[]]ZT[[<]V]XVWY[XSZ[WU      51   \n",
       "...                                                   ...     ...   \n",
       "76239175    VU[TW]UY[TYZU[]]WQZNZ[[[]]ZT[]QZ]]]Z[YXWZ]P]W      52   \n",
       "76239176   VU[TW]UY[TYZU[]]WQZNZ[[[]]ZT[]QZ]]]Z[YXWZ]P]W]      52   \n",
       "76239270   VU[TW]UY[TYZU[]]WQZNZ[[[]]ZT[]Z]]]Z[YXWZ]P]W]]      54   \n",
       "76239550     VU[TW]UY[TYZU[]WQZNZ[[[]]ZT[]Z]]]Z[YXWZ]P]W]      55   \n",
       "76240015       VUTW]UY[TYZ[]WQZNZ[[[]]ZT[]Z]]]Z[YXWZ]PW]V      50   \n",
       "\n",
       "                                                      res_t  \\\n",
       "pos                                                           \n",
       "76142500  ,,,...,,.$,....,,...,,.,..,,.,..,..,.,,,..,,.....   \n",
       "76142565  ,,,...,,,....,,..G,,.,-1g..,,.,..,..G,,,..,,.....   \n",
       "76142572  ,,,..-1C.c,,...-1C.,,...,,.,..,,.,..,...,*,.+1...   \n",
       "76142791  ,,...,,-2at,...*,,...,,.,..-3ATG,,.,..,...,,,....   \n",
       "76142869  ,,...,,,G...,,...,*T,G.,,.,..t...g,,..,,....-1...   \n",
       "...                                                     ...   \n",
       "76239175  ..,,,,.*,,..,.,,,,,..,,....,..,..,,,,+1t.,..C....   \n",
       "76239176  ..,,,,.*,,.-1G.,.,,,,,..,,....,..,..,,,,.+1G,....   \n",
       "76239270  ..,,,,.,,,..,.,,,,,.+1G.,,....,..-2AT,..,,,,.,...   \n",
       "76239550  *A**,*.,,t*.,**,,*..*a..*****,*.,,,.*.*..**,*....   \n",
       "76240015  GA,,,,+2aat,,.*,G,,,..*...,...+1A.,,a.,.+1G..+...   \n",
       "\n",
       "                                                     qual_t  ... SV_t SV_n  \\\n",
       "pos                                                          ...             \n",
       "76142500  38>A479:B7B<44G2@CA+B+475>298@=75D>1<?;3(9@5=?...  ...    2    2   \n",
       "76142565  764J0159<A)./=7(%(;.'$#$%6&)72:.5$'7=(/<7,30./...  ...    2    2   \n",
       "76142572  ?&+=>:+$'@:594=>=.(38$93-$>(A7-:=<$(?9@$.>@<:D...  ...    2    2   \n",
       "76142791  :6,&=?)>>;<(95<0<89:=?&29;<7;<@:5+;996?86<=49?...  ...    2    2   \n",
       "76142869  *99+.)?9/&),<)20&20(+$).'1<=-170/%:7+(%/-*.%<-...  ...    2    2   \n",
       "...                                                     ...  ...  ...  ...   \n",
       "76239175  ';9BB?6$=..-.3:-6?./A<?&()6/9.==+<(&%$=,/#<;F1...  ...    2    2   \n",
       "76239176  -4;@CB<$<8*)+4829A80C=C%()70;099.?+(/$=56%:@:/...  ...    2    2   \n",
       "76239270  %9'9.?%(5&6*,77$,1&%4&'*$''($%>%0<)?&*5*+*<,39...  ...    2    2   \n",
       "76239550  -+0'>?@<7(7C5:;/9+C=57<B7%:61?>=:>&99;)1:4<;=:...  ...    2    2   \n",
       "76240015  %(-$'(&%%$51*+&-6&,)-.'%5/(%4#4-43.39'%'%.',7-...  ...    2    2   \n",
       "\n",
       "                                            res_n_no_carrot  \\\n",
       "pos                                                           \n",
       "76142500      ..*,......,,....,,........,.,.,.,,,,,.,,....,   \n",
       "76142565      ..,,.....,,....,,T.......t.,.,.,,,,,.,,....,,   \n",
       "76142572  .-1C.,,.-1C...-1C.,,....-1C,,.......-1C.,.,.,+...   \n",
       "76142791     ..,,....,,....,,........,.,.,.,,,,,.,,....,c,.   \n",
       "76142869  .*,,-1g..G.,,....,,........,.,.,.,,,,,-1g.,,.....   \n",
       "...                                                     ...   \n",
       "76239175  ,...,.,,,,,,,,A.,.,.,,.+2AA,,A,,....,..,..,,,...,   \n",
       "76239176  ,...,.,t,,,,,,..-2GA,.,.,,.,,.,,...-1G.,..,..,...   \n",
       "76239270  ,...,.+1T,,,*,,,,..,.,.+1T,,.,,.,,...,..,.+1T*...   \n",
       "76239550  ,*.*,**,,,,,,,.**,*,,.,,.,+5ggttt,*.A,**,..,,*...   \n",
       "76240015    ,..,.,,,,,,,.,.,.,t.,,.,+3aag,A..,..,..,t,..,.,   \n",
       "\n",
       "                                            res_t_no_carrot  \\\n",
       "pos                                                           \n",
       "76142500  ,,,...,,.$,....,,...,,.,..,,.,..,..,.,,,..,,.....   \n",
       "76142565  ,,,...,,,....,,..G,,.,-1g..,,.,..,..G,,,..,,.....   \n",
       "76142572  ,,,..-1C.c,,...-1C.,,...,,.,..,,.,..,...,*,.+1...   \n",
       "76142791  ,,...,,-2at,...*,,...,,.,..-3ATG,,.,..,...,,,....   \n",
       "76142869  ,,...,,,G...,,...,*T,G.,,.,..t...g,,..,,....-1...   \n",
       "...                                                     ...   \n",
       "76239175  ..,,,,.*,,..,.,,,,,..,,....,..,..,,,,+1t.,..C....   \n",
       "76239176  ..,,,,.*,,.-1G.,.,,,,,..,,....,..,..,,,,.+1G,....   \n",
       "76239270  ..,,,,.,,,..,.,,,,,.+1G.,,....,..-2AT,..,,,,.,...   \n",
       "76239550  *A**,*.,,t*.,**,,*..*a..*****,*.,,,.*.*..**,*....   \n",
       "76240015  GA,,,,+2aat,,.*,G,,,..*...,...+1A.,,a.,.+1G..+...   \n",
       "\n",
       "                                           t_indel_free_res             t_ins  \\\n",
       "pos                                                                             \n",
       "76142500  ,,,...,,.$,....,,...,,.,..,,.,..,..,.,,,..,,.....                []   \n",
       "76142565  ,,,...,,,....,,..G,,.,..,,.,..,..G,,,..,,........                []   \n",
       "76142572  ,,,...c,,....,,...,,.,..,,.,..,...,*,..,,........               [C]   \n",
       "76142791  ,,...,,,...*,,...,,.,..,,.,..,...,,,..,,.....,...                []   \n",
       "76142869  ,,...,,,G...,,...,*T,G.,,.,..t...g,,..,,.....g...                []   \n",
       "...                                                     ...               ...   \n",
       "76239175  ..,,,,.*,,..,.,,,,,..,,....,..,..,,,,.,..C.,.,...               [t]   \n",
       "76239176  ..,,,,.*,,..,.,,,,,..,,....,..,..,,,,.,....,.,...               [G]   \n",
       "76239270  ..,,,,.,,,..,.,,,,,..,,....,..,..,,,,.,.**.,.,...               [G]   \n",
       "76239550  *A**,*.,,t*.,**,,*..*a..*****,*.,,,.*.*..**,*....                []   \n",
       "76240015  GA,,,,t,,.*,G,,,..*...,....,,a.,.....,gAT..,.....  [aa, A, G, G, G]   \n",
       "\n",
       "                   t_dels                                 n_indel_free_res  \\\n",
       "pos                                                                          \n",
       "76142500               []    ..*,......,,....,,........,.,.,.,,,,,.,,....,   \n",
       "76142565              [g]    ..,,.....,,....,,T.......t.,.,.,,,,,.,,....,,   \n",
       "76142572  [C, C, C, C, C]   ..,,.....,,....,,........,.,.,.,*,**.,*....,,,   \n",
       "76142791        [at, ATG]   ..,,....,,....,,........,.,.,.,,,,,.,,....,c,.   \n",
       "76142869              [G]  .*,,..G.,,....,,........,.,.,.,,,,,.,,....,,,..   \n",
       "...                   ...                                              ...   \n",
       "76239175               []    ,...,.,,,,,,,,A.,.,.,,.,,A,,....,..,..,,,...,   \n",
       "76239176              [G]   ,...,.,t,,,,,,..,.,.,,.,,.,,....,..,..,,,...,,   \n",
       "76239270       [AT, A, A]   ,...,.,,,*,,,,..,.,.,,.,,.,,...,..,.*,,t...,,,   \n",
       "76239550               []     ,*.*,**,,,,,,,.**,*,,.,,.,,*.A,**,..,,*.*.,.   \n",
       "76240015               []       ,..,.,,,,,,,.,.,.,t.,,.,,A..,..,..,t,..,.,   \n",
       "\n",
       "              n_ins                 n_dels  \n",
       "pos                                         \n",
       "76142500         []                     []  \n",
       "76142565         []                     []  \n",
       "76142572        [c]  [C, C, C, C, C, C, c]  \n",
       "76142791         []                     []  \n",
       "76142869         []                 [g, g]  \n",
       "...             ...                    ...  \n",
       "76239175       [AA]                     []  \n",
       "76239176         []                [GA, G]  \n",
       "76239270  [T, T, T]                    [A]  \n",
       "76239550    [ggttt]                     []  \n",
       "76240015      [aag]                     []  \n",
       "\n",
       "[759 rows x 26 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#wtf!! .-1C.,,.-1C...-1C.,,....-1C,,.......-1C.,.,.,+...    @76142572 why + by itself?? ... is truncation... ffs\n",
    "df[df.res_n.str.contains('$', regex=False)]#.to_csv('~/Downloads/dola.csv')\n",
    "df[df.res_n.str.contains('^', regex=False)]#.to_csv('~/Downloads/carrot.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\n",
    "    path,\n",
    "    dtype={\n",
    "        DataSchema.AMOUNT: float,\n",
    "        DataSchema.CATEGORY: str,\n",
    "        DataSchema.DATE: str,\n",
    "    },\n",
    "    parse_dates=[DataSchema.DATE],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>size</th>\n",
       "      <th>size_kb</th>\n",
       "      <th>size_mb</th>\n",
       "      <th>size_gb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EBBZP</td>\n",
       "      <td>9237935</td>\n",
       "      <td>9021.42</td>\n",
       "      <td>8.81</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A7HC2</td>\n",
       "      <td>3127235</td>\n",
       "      <td>3053.94</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UTJ34</td>\n",
       "      <td>6069830</td>\n",
       "      <td>5927.57</td>\n",
       "      <td>5.79</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTLJ9</td>\n",
       "      <td>15951978</td>\n",
       "      <td>15578.10</td>\n",
       "      <td>15.21</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CELYH</td>\n",
       "      <td>8556300</td>\n",
       "      <td>8355.76</td>\n",
       "      <td>8.16</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>5UM50</td>\n",
       "      <td>10262913</td>\n",
       "      <td>10022.38</td>\n",
       "      <td>9.79</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>FFHPH</td>\n",
       "      <td>11270724</td>\n",
       "      <td>11006.57</td>\n",
       "      <td>10.75</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>3WJH2</td>\n",
       "      <td>13253839</td>\n",
       "      <td>12943.20</td>\n",
       "      <td>12.64</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>89ZJE</td>\n",
       "      <td>16825160</td>\n",
       "      <td>16430.82</td>\n",
       "      <td>16.05</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>PDR1L</td>\n",
       "      <td>7449552</td>\n",
       "      <td>7274.95</td>\n",
       "      <td>7.10</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         name      size   size_kb  size_mb  size_gb\n",
       "0       EBBZP   9237935   9021.42     8.81     0.01\n",
       "1       A7HC2   3127235   3053.94     2.98     0.00\n",
       "2       UTJ34   6069830   5927.57     5.79     0.01\n",
       "3       CTLJ9  15951978  15578.10    15.21     0.01\n",
       "4       CELYH   8556300   8355.76     8.16     0.01\n",
       "...       ...       ...       ...      ...      ...\n",
       "999995  5UM50  10262913  10022.38     9.79     0.01\n",
       "999996  FFHPH  11270724  11006.57    10.75     0.01\n",
       "999997  3WJH2  13253839  12943.20    12.64     0.01\n",
       "999998  89ZJE  16825160  16430.82    16.05     0.02\n",
       "999999  PDR1L   7449552   7274.95     7.10     0.01\n",
       "\n",
       "[1000000 rows x 5 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import locale\n",
    "import random, string\n",
    "\n",
    "N = 1000000\n",
    "np.random.seed(0)\n",
    "rand_name_lst=[''.join(random.choices(string.ascii_uppercase + string.digits, k=5)) \\\n",
    "                            for i in range(N)]\n",
    "rand_int_lst = np.random.randint(912131,18293293,size=N)\n",
    "df = pd.DataFrame({'name': rand_name_lst , 'size': rand_int_lst})\n",
    "\n",
    "df.sample(5)\n",
    "\n",
    "def convert_size(x):\n",
    "    kb = round((x / 1024.0),2)\n",
    "    mb = round((x / (1024.0 ** 2)),2)\n",
    "    gb = round((x / (1024 **3)),2)\n",
    "\n",
    "    return pd.Series([kb, mb, gb])\n",
    "df[['size_kb','size_mb','size_gb']]=df['size'].apply(convert_size)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>size</th>\n",
       "      <th>size_kb</th>\n",
       "      <th>size_mb</th>\n",
       "      <th>size_gb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EBBZP</td>\n",
       "      <td>9237935</td>\n",
       "      <td>9021.42</td>\n",
       "      <td>8.81</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A7HC2</td>\n",
       "      <td>3127235</td>\n",
       "      <td>3053.94</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UTJ34</td>\n",
       "      <td>6069830</td>\n",
       "      <td>5927.57</td>\n",
       "      <td>5.79</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTLJ9</td>\n",
       "      <td>15951978</td>\n",
       "      <td>15578.10</td>\n",
       "      <td>15.21</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CELYH</td>\n",
       "      <td>8556300</td>\n",
       "      <td>8355.76</td>\n",
       "      <td>8.16</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999995</th>\n",
       "      <td>5UM50</td>\n",
       "      <td>10262913</td>\n",
       "      <td>10022.38</td>\n",
       "      <td>9.79</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999996</th>\n",
       "      <td>FFHPH</td>\n",
       "      <td>11270724</td>\n",
       "      <td>11006.57</td>\n",
       "      <td>10.75</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999997</th>\n",
       "      <td>3WJH2</td>\n",
       "      <td>13253839</td>\n",
       "      <td>12943.20</td>\n",
       "      <td>12.64</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999998</th>\n",
       "      <td>89ZJE</td>\n",
       "      <td>16825160</td>\n",
       "      <td>16430.82</td>\n",
       "      <td>16.05</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999999</th>\n",
       "      <td>PDR1L</td>\n",
       "      <td>7449552</td>\n",
       "      <td>7274.95</td>\n",
       "      <td>7.10</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000000 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         name      size   size_kb  size_mb  size_gb\n",
       "0       EBBZP   9237935   9021.42     8.81     0.01\n",
       "1       A7HC2   3127235   3053.94     2.98     0.00\n",
       "2       UTJ34   6069830   5927.57     5.79     0.01\n",
       "3       CTLJ9  15951978  15578.10    15.21     0.01\n",
       "4       CELYH   8556300   8355.76     8.16     0.01\n",
       "...       ...       ...       ...      ...      ...\n",
       "999995  5UM50  10262913  10022.38     9.79     0.01\n",
       "999996  FFHPH  11270724  11006.57    10.75     0.01\n",
       "999997  3WJH2  13253839  12943.20    12.64     0.01\n",
       "999998  89ZJE  16825160  16430.82    16.05     0.02\n",
       "999999  PDR1L   7449552   7274.95     7.10     0.01\n",
       "\n",
       "[1000000 rows x 5 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['size_kb'],df['size_mb'],df['size_gb']=convert_size(df['size'])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([], ['-1C', '-1C', '-1C', '-1C'])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "txt = \".-1C.,,.-1C...-1C.,,....-1C,,..\"\n",
    "re.findall(\"\\+[0-9]+[ACGTNacgtn]+\", txt), re.findall('\\-[0-9]+[ACGTNacgtn]+', txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'..,,.....,,....,,..'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for indel in cc.keys():\n",
    "    txt = ''.join(txt.split(indel))\n",
    "txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s1 ='CPPffff'\n",
    "s2='gggg'\n",
    "\n",
    "any(map(s1.startswith, ['CPP', 'CGD', 'CPG', 'CNE', 'CNL']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on PlotAccessor in module pandas.plotting._core object:\n",
      "\n",
      "class PlotAccessor(pandas.core.base.PandasObject)\n",
      " |  PlotAccessor(data)\n",
      " |  \n",
      " |  Make plots of Series or DataFrame.\n",
      " |  \n",
      " |  Uses the backend specified by the\n",
      " |  option ``plotting.backend``. By default, matplotlib is used.\n",
      " |  \n",
      " |  Parameters\n",
      " |  ----------\n",
      " |  data : Series or DataFrame\n",
      " |      The object for which the method is called.\n",
      " |  x : label or position, default None\n",
      " |      Only used if data is a DataFrame.\n",
      " |  y : label, position or list of label, positions, default None\n",
      " |      Allows plotting of one column versus another. Only used if data is a\n",
      " |      DataFrame.\n",
      " |  kind : str\n",
      " |      The kind of plot to produce:\n",
      " |  \n",
      " |      - 'line' : line plot (default)\n",
      " |      - 'bar' : vertical bar plot\n",
      " |      - 'barh' : horizontal bar plot\n",
      " |      - 'hist' : histogram\n",
      " |      - 'box' : boxplot\n",
      " |      - 'kde' : Kernel Density Estimation plot\n",
      " |      - 'density' : same as 'kde'\n",
      " |      - 'area' : area plot\n",
      " |      - 'pie' : pie plot\n",
      " |      - 'scatter' : scatter plot (DataFrame only)\n",
      " |      - 'hexbin' : hexbin plot (DataFrame only)\n",
      " |  ax : matplotlib axes object, default None\n",
      " |      An axes of the current figure.\n",
      " |  subplots : bool, default False\n",
      " |      Make separate subplots for each column.\n",
      " |  sharex : bool, default True if ax is None else False\n",
      " |      In case ``subplots=True``, share x axis and set some x axis labels\n",
      " |      to invisible; defaults to True if ax is None otherwise False if\n",
      " |      an ax is passed in; Be aware, that passing in both an ax and\n",
      " |      ``sharex=True`` will alter all x axis labels for all axis in a figure.\n",
      " |  sharey : bool, default False\n",
      " |      In case ``subplots=True``, share y axis and set some y axis labels to invisible.\n",
      " |  layout : tuple, optional\n",
      " |      (rows, columns) for the layout of subplots.\n",
      " |  figsize : a tuple (width, height) in inches\n",
      " |      Size of a figure object.\n",
      " |  use_index : bool, default True\n",
      " |      Use index as ticks for x axis.\n",
      " |  title : str or list\n",
      " |      Title to use for the plot. If a string is passed, print the string\n",
      " |      at the top of the figure. If a list is passed and `subplots` is\n",
      " |      True, print each item in the list above the corresponding subplot.\n",
      " |  grid : bool, default None (matlab style default)\n",
      " |      Axis grid lines.\n",
      " |  legend : bool or {'reverse'}\n",
      " |      Place legend on axis subplots.\n",
      " |  style : list or dict\n",
      " |      The matplotlib line style per column.\n",
      " |  logx : bool or 'sym', default False\n",
      " |      Use log scaling or symlog scaling on x axis.\n",
      " |      .. versionchanged:: 0.25.0\n",
      " |  \n",
      " |  logy : bool or 'sym' default False\n",
      " |      Use log scaling or symlog scaling on y axis.\n",
      " |      .. versionchanged:: 0.25.0\n",
      " |  \n",
      " |  loglog : bool or 'sym', default False\n",
      " |      Use log scaling or symlog scaling on both x and y axes.\n",
      " |      .. versionchanged:: 0.25.0\n",
      " |  \n",
      " |  xticks : sequence\n",
      " |      Values to use for the xticks.\n",
      " |  yticks : sequence\n",
      " |      Values to use for the yticks.\n",
      " |  xlim : 2-tuple/list\n",
      " |      Set the x limits of the current axes.\n",
      " |  ylim : 2-tuple/list\n",
      " |      Set the y limits of the current axes.\n",
      " |  xlabel : label, optional\n",
      " |      Name to use for the xlabel on x-axis. Default uses index name as xlabel, or the\n",
      " |      x-column name for planar plots.\n",
      " |  \n",
      " |      .. versionadded:: 1.1.0\n",
      " |  \n",
      " |      .. versionchanged:: 1.2.0\n",
      " |  \n",
      " |         Now applicable to planar plots (`scatter`, `hexbin`).\n",
      " |  \n",
      " |  ylabel : label, optional\n",
      " |      Name to use for the ylabel on y-axis. Default will show no ylabel, or the\n",
      " |      y-column name for planar plots.\n",
      " |  \n",
      " |      .. versionadded:: 1.1.0\n",
      " |  \n",
      " |      .. versionchanged:: 1.2.0\n",
      " |  \n",
      " |         Now applicable to planar plots (`scatter`, `hexbin`).\n",
      " |  \n",
      " |  rot : int, default None\n",
      " |      Rotation for ticks (xticks for vertical, yticks for horizontal\n",
      " |      plots).\n",
      " |  fontsize : int, default None\n",
      " |      Font size for xticks and yticks.\n",
      " |  colormap : str or matplotlib colormap object, default None\n",
      " |      Colormap to select colors from. If string, load colormap with that\n",
      " |      name from matplotlib.\n",
      " |  colorbar : bool, optional\n",
      " |      If True, plot colorbar (only relevant for 'scatter' and 'hexbin'\n",
      " |      plots).\n",
      " |  position : float\n",
      " |      Specify relative alignments for bar plot layout.\n",
      " |      From 0 (left/bottom-end) to 1 (right/top-end). Default is 0.5\n",
      " |      (center).\n",
      " |  table : bool, Series or DataFrame, default False\n",
      " |      If True, draw a table using the data in the DataFrame and the data\n",
      " |      will be transposed to meet matplotlib's default layout.\n",
      " |      If a Series or DataFrame is passed, use passed data to draw a\n",
      " |      table.\n",
      " |  yerr : DataFrame, Series, array-like, dict and str\n",
      " |      See :ref:`Plotting with Error Bars <visualization.errorbars>` for\n",
      " |      detail.\n",
      " |  xerr : DataFrame, Series, array-like, dict and str\n",
      " |      Equivalent to yerr.\n",
      " |  stacked : bool, default False in line and bar plots, and True in area plot\n",
      " |      If True, create stacked plot.\n",
      " |  sort_columns : bool, default False\n",
      " |      Sort column names to determine plot ordering.\n",
      " |  secondary_y : bool or sequence, default False\n",
      " |      Whether to plot on the secondary y-axis if a list/tuple, which\n",
      " |      columns to plot on secondary y-axis.\n",
      " |  mark_right : bool, default True\n",
      " |      When using a secondary_y axis, automatically mark the column\n",
      " |      labels with \"(right)\" in the legend.\n",
      " |  include_bool : bool, default is False\n",
      " |      If True, boolean values can be plotted.\n",
      " |  backend : str, default None\n",
      " |      Backend to use instead of the backend specified in the option\n",
      " |      ``plotting.backend``. For instance, 'matplotlib'. Alternatively, to\n",
      " |      specify the ``plotting.backend`` for the whole session, set\n",
      " |      ``pd.options.plotting.backend``.\n",
      " |  \n",
      " |      .. versionadded:: 1.0.0\n",
      " |  \n",
      " |  **kwargs\n",
      " |      Options to pass to matplotlib plotting method.\n",
      " |  \n",
      " |  Returns\n",
      " |  -------\n",
      " |  :class:`matplotlib.axes.Axes` or numpy.ndarray of them\n",
      " |      If the backend is not the default matplotlib one, the return value\n",
      " |      will be the object returned by the backend.\n",
      " |  \n",
      " |  Notes\n",
      " |  -----\n",
      " |  - See matplotlib documentation online for more on this subject\n",
      " |  - If `kind` = 'bar' or 'barh', you can specify relative alignments\n",
      " |    for bar plot layout by `position` keyword.\n",
      " |    From 0 (left/bottom-end) to 1 (right/top-end). Default is 0.5\n",
      " |    (center)\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      PlotAccessor\n",
      " |      pandas.core.base.PandasObject\n",
      " |      pandas.core.accessor.DirNamesMixin\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __call__(self, *args, **kwargs)\n",
      " |      Make plots of Series or DataFrame.\n",
      " |      \n",
      " |      Uses the backend specified by the\n",
      " |      option ``plotting.backend``. By default, matplotlib is used.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : Series or DataFrame\n",
      " |          The object for which the method is called.\n",
      " |      x : label or position, default None\n",
      " |          Only used if data is a DataFrame.\n",
      " |      y : label, position or list of label, positions, default None\n",
      " |          Allows plotting of one column versus another. Only used if data is a\n",
      " |          DataFrame.\n",
      " |      kind : str\n",
      " |          The kind of plot to produce:\n",
      " |      \n",
      " |          - 'line' : line plot (default)\n",
      " |          - 'bar' : vertical bar plot\n",
      " |          - 'barh' : horizontal bar plot\n",
      " |          - 'hist' : histogram\n",
      " |          - 'box' : boxplot\n",
      " |          - 'kde' : Kernel Density Estimation plot\n",
      " |          - 'density' : same as 'kde'\n",
      " |          - 'area' : area plot\n",
      " |          - 'pie' : pie plot\n",
      " |          - 'scatter' : scatter plot (DataFrame only)\n",
      " |          - 'hexbin' : hexbin plot (DataFrame only)\n",
      " |      ax : matplotlib axes object, default None\n",
      " |          An axes of the current figure.\n",
      " |      subplots : bool, default False\n",
      " |          Make separate subplots for each column.\n",
      " |      sharex : bool, default True if ax is None else False\n",
      " |          In case ``subplots=True``, share x axis and set some x axis labels\n",
      " |          to invisible; defaults to True if ax is None otherwise False if\n",
      " |          an ax is passed in; Be aware, that passing in both an ax and\n",
      " |          ``sharex=True`` will alter all x axis labels for all axis in a figure.\n",
      " |      sharey : bool, default False\n",
      " |          In case ``subplots=True``, share y axis and set some y axis labels to invisible.\n",
      " |      layout : tuple, optional\n",
      " |          (rows, columns) for the layout of subplots.\n",
      " |      figsize : a tuple (width, height) in inches\n",
      " |          Size of a figure object.\n",
      " |      use_index : bool, default True\n",
      " |          Use index as ticks for x axis.\n",
      " |      title : str or list\n",
      " |          Title to use for the plot. If a string is passed, print the string\n",
      " |          at the top of the figure. If a list is passed and `subplots` is\n",
      " |          True, print each item in the list above the corresponding subplot.\n",
      " |      grid : bool, default None (matlab style default)\n",
      " |          Axis grid lines.\n",
      " |      legend : bool or {'reverse'}\n",
      " |          Place legend on axis subplots.\n",
      " |      style : list or dict\n",
      " |          The matplotlib line style per column.\n",
      " |      logx : bool or 'sym', default False\n",
      " |          Use log scaling or symlog scaling on x axis.\n",
      " |          .. versionchanged:: 0.25.0\n",
      " |      \n",
      " |      logy : bool or 'sym' default False\n",
      " |          Use log scaling or symlog scaling on y axis.\n",
      " |          .. versionchanged:: 0.25.0\n",
      " |      \n",
      " |      loglog : bool or 'sym', default False\n",
      " |          Use log scaling or symlog scaling on both x and y axes.\n",
      " |          .. versionchanged:: 0.25.0\n",
      " |      \n",
      " |      xticks : sequence\n",
      " |          Values to use for the xticks.\n",
      " |      yticks : sequence\n",
      " |          Values to use for the yticks.\n",
      " |      xlim : 2-tuple/list\n",
      " |          Set the x limits of the current axes.\n",
      " |      ylim : 2-tuple/list\n",
      " |          Set the y limits of the current axes.\n",
      " |      xlabel : label, optional\n",
      " |          Name to use for the xlabel on x-axis. Default uses index name as xlabel, or the\n",
      " |          x-column name for planar plots.\n",
      " |      \n",
      " |          .. versionadded:: 1.1.0\n",
      " |      \n",
      " |          .. versionchanged:: 1.2.0\n",
      " |      \n",
      " |             Now applicable to planar plots (`scatter`, `hexbin`).\n",
      " |      \n",
      " |      ylabel : label, optional\n",
      " |          Name to use for the ylabel on y-axis. Default will show no ylabel, or the\n",
      " |          y-column name for planar plots.\n",
      " |      \n",
      " |          .. versionadded:: 1.1.0\n",
      " |      \n",
      " |          .. versionchanged:: 1.2.0\n",
      " |      \n",
      " |             Now applicable to planar plots (`scatter`, `hexbin`).\n",
      " |      \n",
      " |      rot : int, default None\n",
      " |          Rotation for ticks (xticks for vertical, yticks for horizontal\n",
      " |          plots).\n",
      " |      fontsize : int, default None\n",
      " |          Font size for xticks and yticks.\n",
      " |      colormap : str or matplotlib colormap object, default None\n",
      " |          Colormap to select colors from. If string, load colormap with that\n",
      " |          name from matplotlib.\n",
      " |      colorbar : bool, optional\n",
      " |          If True, plot colorbar (only relevant for 'scatter' and 'hexbin'\n",
      " |          plots).\n",
      " |      position : float\n",
      " |          Specify relative alignments for bar plot layout.\n",
      " |          From 0 (left/bottom-end) to 1 (right/top-end). Default is 0.5\n",
      " |          (center).\n",
      " |      table : bool, Series or DataFrame, default False\n",
      " |          If True, draw a table using the data in the DataFrame and the data\n",
      " |          will be transposed to meet matplotlib's default layout.\n",
      " |          If a Series or DataFrame is passed, use passed data to draw a\n",
      " |          table.\n",
      " |      yerr : DataFrame, Series, array-like, dict and str\n",
      " |          See :ref:`Plotting with Error Bars <visualization.errorbars>` for\n",
      " |          detail.\n",
      " |      xerr : DataFrame, Series, array-like, dict and str\n",
      " |          Equivalent to yerr.\n",
      " |      stacked : bool, default False in line and bar plots, and True in area plot\n",
      " |          If True, create stacked plot.\n",
      " |      sort_columns : bool, default False\n",
      " |          Sort column names to determine plot ordering.\n",
      " |      secondary_y : bool or sequence, default False\n",
      " |          Whether to plot on the secondary y-axis if a list/tuple, which\n",
      " |          columns to plot on secondary y-axis.\n",
      " |      mark_right : bool, default True\n",
      " |          When using a secondary_y axis, automatically mark the column\n",
      " |          labels with \"(right)\" in the legend.\n",
      " |      include_bool : bool, default is False\n",
      " |          If True, boolean values can be plotted.\n",
      " |      backend : str, default None\n",
      " |          Backend to use instead of the backend specified in the option\n",
      " |          ``plotting.backend``. For instance, 'matplotlib'. Alternatively, to\n",
      " |          specify the ``plotting.backend`` for the whole session, set\n",
      " |          ``pd.options.plotting.backend``.\n",
      " |      \n",
      " |          .. versionadded:: 1.0.0\n",
      " |      \n",
      " |      **kwargs\n",
      " |          Options to pass to matplotlib plotting method.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`matplotlib.axes.Axes` or numpy.ndarray of them\n",
      " |          If the backend is not the default matplotlib one, the return value\n",
      " |          will be the object returned by the backend.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      - See matplotlib documentation online for more on this subject\n",
      " |      - If `kind` = 'bar' or 'barh', you can specify relative alignments\n",
      " |        for bar plot layout by `position` keyword.\n",
      " |        From 0 (left/bottom-end) to 1 (right/top-end). Default is 0.5\n",
      " |        (center)\n",
      " |  \n",
      " |  __init__(self, data)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  area(self, x=None, y=None, **kwargs)\n",
      " |      Draw a stacked area plot.\n",
      " |      \n",
      " |      An area plot displays quantitative data visually.\n",
      " |      This function wraps the matplotlib area function.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      x : label or position, optional\n",
      " |          Coordinates for the X axis. By default uses the index.\n",
      " |      y : label or position, optional\n",
      " |          Column to plot. By default uses all columns.\n",
      " |      stacked : bool, default True\n",
      " |          Area plots are stacked by default. Set to False to create a\n",
      " |          unstacked plot.\n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.axes.Axes or numpy.ndarray\n",
      " |          Area plot, or array of area plots if subplots is True.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      DataFrame.plot : Make plots of DataFrame using matplotlib / pylab.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Draw an area plot based on basic business metrics:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> df = pd.DataFrame({\n",
      " |          ...     'sales': [3, 2, 3, 9, 10, 6],\n",
      " |          ...     'signups': [5, 5, 6, 12, 14, 13],\n",
      " |          ...     'visits': [20, 42, 28, 62, 81, 50],\n",
      " |          ... }, index=pd.date_range(start='2018/01/01', end='2018/07/01',\n",
      " |          ...                        freq='M'))\n",
      " |          >>> ax = df.plot.area()\n",
      " |      \n",
      " |      Area plots are stacked by default. To produce an unstacked plot,\n",
      " |      pass ``stacked=False``:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = df.plot.area(stacked=False)\n",
      " |      \n",
      " |      Draw an area plot for a single column:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = df.plot.area(y='sales')\n",
      " |      \n",
      " |      Draw with a different `x`:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> df = pd.DataFrame({\n",
      " |          ...     'sales': [3, 2, 3],\n",
      " |          ...     'visits': [20, 42, 28],\n",
      " |          ...     'day': [1, 2, 3],\n",
      " |          ... })\n",
      " |          >>> ax = df.plot.area(x='day')\n",
      " |  \n",
      " |  bar(self, x=None, y=None, **kwargs)\n",
      " |      Vertical bar plot.\n",
      " |      \n",
      " |      A bar plot is a plot that presents categorical data with\n",
      " |      rectangular bars with lengths proportional to the values that they\n",
      " |      represent. A bar plot shows comparisons among discrete categories. One\n",
      " |      axis of the plot shows the specific categories being compared, and the\n",
      " |      other axis represents a measured value.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      x : label or position, optional\n",
      " |          Allows plotting of one column versus another. If not specified,\n",
      " |          the index of the DataFrame is used.\n",
      " |      y : label or position, optional\n",
      " |          Allows plotting of one column versus another. If not specified,\n",
      " |          all numerical columns are used.\n",
      " |      color : str, array-like, or dict, optional\n",
      " |          The color for each of the DataFrame's columns. Possible values are:\n",
      " |      \n",
      " |          - A single color string referred to by name, RGB or RGBA code,\n",
      " |              for instance 'red' or '#a98d19'.\n",
      " |      \n",
      " |          - A sequence of color strings referred to by name, RGB or RGBA\n",
      " |              code, which will be used for each column recursively. For\n",
      " |              instance ['green','yellow'] each column's bar will be filled in\n",
      " |              green or yellow, alternatively. If there is only a single column to\n",
      " |              be plotted, then only the first color from the color list will be\n",
      " |              used.\n",
      " |      \n",
      " |          - A dict of the form {column name : color}, so that each column will be\n",
      " |              colored accordingly. For example, if your columns are called `a` and\n",
      " |              `b`, then passing {'a': 'green', 'b': 'red'} will color bars for\n",
      " |              column `a` in green and bars for column `b` in red.\n",
      " |      \n",
      " |          .. versionadded:: 1.1.0\n",
      " |      \n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.axes.Axes or np.ndarray of them\n",
      " |          An ndarray is returned with one :class:`matplotlib.axes.Axes`\n",
      " |          per column when ``subplots=True``.\n",
      " |      \n",
      " |              See Also\n",
      " |              --------\n",
      " |              DataFrame.plot.barh : Horizontal bar plot.\n",
      " |              DataFrame.plot : Make plots of a DataFrame.\n",
      " |              matplotlib.pyplot.bar : Make a bar plot with matplotlib.\n",
      " |      \n",
      " |              Examples\n",
      " |              --------\n",
      " |              Basic plot.\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> df = pd.DataFrame({'lab':['A', 'B', 'C'], 'val':[10, 30, 20]})\n",
      " |                  >>> ax = df.plot.bar(x='lab', y='val', rot=0)\n",
      " |      \n",
      " |              Plot a whole dataframe to a bar plot. Each column is assigned a\n",
      " |              distinct color, and each row is nested in a group along the\n",
      " |              horizontal axis.\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> speed = [0.1, 17.5, 40, 48, 52, 69, 88]\n",
      " |                  >>> lifespan = [2, 8, 70, 1.5, 25, 12, 28]\n",
      " |                  >>> index = ['snail', 'pig', 'elephant',\n",
      " |                  ...          'rabbit', 'giraffe', 'coyote', 'horse']\n",
      " |                  >>> df = pd.DataFrame({'speed': speed,\n",
      " |                  ...                    'lifespan': lifespan}, index=index)\n",
      " |                  >>> ax = df.plot.bar(rot=0)\n",
      " |      \n",
      " |              Plot stacked bar charts for the DataFrame\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> ax = df.plot.bar(stacked=True)\n",
      " |      \n",
      " |              Instead of nesting, the figure can be split by column with\n",
      " |              ``subplots=True``. In this case, a :class:`numpy.ndarray` of\n",
      " |              :class:`matplotlib.axes.Axes` are returned.\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> axes = df.plot.bar(rot=0, subplots=True)\n",
      " |                  >>> axes[1].legend(loc=2)  # doctest: +SKIP\n",
      " |      \n",
      " |              If you don't like the default colours, you can specify how you'd\n",
      " |              like each column to be colored.\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> axes = df.plot.bar(\n",
      " |                  ...     rot=0, subplots=True, color={\"speed\": \"red\", \"lifespan\": \"green\"}\n",
      " |                  ... )\n",
      " |                  >>> axes[1].legend(loc=2)  # doctest: +SKIP\n",
      " |      \n",
      " |              Plot a single column.\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> ax = df.plot.bar(y='speed', rot=0)\n",
      " |      \n",
      " |              Plot only selected categories for the DataFrame.\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> ax = df.plot.bar(x='lifespan', rot=0)\n",
      " |  \n",
      " |  barh(self, x=None, y=None, **kwargs)\n",
      " |      Make a horizontal bar plot.\n",
      " |      \n",
      " |      A horizontal bar plot is a plot that presents quantitative data with\n",
      " |      rectangular bars with lengths proportional to the values that they\n",
      " |      represent. A bar plot shows comparisons among discrete categories. One\n",
      " |      axis of the plot shows the specific categories being compared, and the\n",
      " |      other axis represents a measured value.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      x : label or position, optional\n",
      " |          Allows plotting of one column versus another. If not specified,\n",
      " |          the index of the DataFrame is used.\n",
      " |      y : label or position, optional\n",
      " |          Allows plotting of one column versus another. If not specified,\n",
      " |          all numerical columns are used.\n",
      " |      color : str, array-like, or dict, optional\n",
      " |          The color for each of the DataFrame's columns. Possible values are:\n",
      " |      \n",
      " |          - A single color string referred to by name, RGB or RGBA code,\n",
      " |              for instance 'red' or '#a98d19'.\n",
      " |      \n",
      " |          - A sequence of color strings referred to by name, RGB or RGBA\n",
      " |              code, which will be used for each column recursively. For\n",
      " |              instance ['green','yellow'] each column's bar will be filled in\n",
      " |              green or yellow, alternatively. If there is only a single column to\n",
      " |              be plotted, then only the first color from the color list will be\n",
      " |              used.\n",
      " |      \n",
      " |          - A dict of the form {column name : color}, so that each column will be\n",
      " |              colored accordingly. For example, if your columns are called `a` and\n",
      " |              `b`, then passing {'a': 'green', 'b': 'red'} will color bars for\n",
      " |              column `a` in green and bars for column `b` in red.\n",
      " |      \n",
      " |          .. versionadded:: 1.1.0\n",
      " |      \n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.axes.Axes or np.ndarray of them\n",
      " |          An ndarray is returned with one :class:`matplotlib.axes.Axes`\n",
      " |          per column when ``subplots=True``.\n",
      " |      \n",
      " |              See Also\n",
      " |              --------\n",
      " |              DataFrame.plot.bar: Vertical bar plot.\n",
      " |              DataFrame.plot : Make plots of DataFrame using matplotlib.\n",
      " |              matplotlib.axes.Axes.bar : Plot a vertical bar plot using matplotlib.\n",
      " |      \n",
      " |              Examples\n",
      " |              --------\n",
      " |              Basic example\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> df = pd.DataFrame({'lab': ['A', 'B', 'C'], 'val': [10, 30, 20]})\n",
      " |                  >>> ax = df.plot.barh(x='lab', y='val')\n",
      " |      \n",
      " |              Plot a whole DataFrame to a horizontal bar plot\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> speed = [0.1, 17.5, 40, 48, 52, 69, 88]\n",
      " |                  >>> lifespan = [2, 8, 70, 1.5, 25, 12, 28]\n",
      " |                  >>> index = ['snail', 'pig', 'elephant',\n",
      " |                  ...          'rabbit', 'giraffe', 'coyote', 'horse']\n",
      " |                  >>> df = pd.DataFrame({'speed': speed,\n",
      " |                  ...                    'lifespan': lifespan}, index=index)\n",
      " |                  >>> ax = df.plot.barh()\n",
      " |      \n",
      " |              Plot stacked barh charts for the DataFrame\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> ax = df.plot.barh(stacked=True)\n",
      " |      \n",
      " |              We can specify colors for each column\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> ax = df.plot.barh(color={\"speed\": \"red\", \"lifespan\": \"green\"})\n",
      " |      \n",
      " |              Plot a column of the DataFrame to a horizontal bar plot\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> speed = [0.1, 17.5, 40, 48, 52, 69, 88]\n",
      " |                  >>> lifespan = [2, 8, 70, 1.5, 25, 12, 28]\n",
      " |                  >>> index = ['snail', 'pig', 'elephant',\n",
      " |                  ...          'rabbit', 'giraffe', 'coyote', 'horse']\n",
      " |                  >>> df = pd.DataFrame({'speed': speed,\n",
      " |                  ...                    'lifespan': lifespan}, index=index)\n",
      " |                  >>> ax = df.plot.barh(y='speed')\n",
      " |      \n",
      " |              Plot DataFrame versus the desired column\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> speed = [0.1, 17.5, 40, 48, 52, 69, 88]\n",
      " |                  >>> lifespan = [2, 8, 70, 1.5, 25, 12, 28]\n",
      " |                  >>> index = ['snail', 'pig', 'elephant',\n",
      " |                  ...          'rabbit', 'giraffe', 'coyote', 'horse']\n",
      " |                  >>> df = pd.DataFrame({'speed': speed,\n",
      " |                  ...                    'lifespan': lifespan}, index=index)\n",
      " |                  >>> ax = df.plot.barh(x='lifespan')\n",
      " |  \n",
      " |  box(self, by=None, **kwargs)\n",
      " |      Make a box plot of the DataFrame columns.\n",
      " |      \n",
      " |      A box plot is a method for graphically depicting groups of numerical\n",
      " |      data through their quartiles.\n",
      " |      The box extends from the Q1 to Q3 quartile values of the data,\n",
      " |      with a line at the median (Q2). The whiskers extend from the edges\n",
      " |      of box to show the range of the data. The position of the whiskers\n",
      " |      is set by default to 1.5*IQR (IQR = Q3 - Q1) from the edges of the\n",
      " |      box. Outlier points are those past the end of the whiskers.\n",
      " |      \n",
      " |      For further details see Wikipedia's\n",
      " |      entry for `boxplot <https://en.wikipedia.org/wiki/Box_plot>`__.\n",
      " |      \n",
      " |      A consideration when using this chart is that the box and the whiskers\n",
      " |      can overlap, which is very common when plotting small sets of data.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      by : str or sequence\n",
      " |          Column in the DataFrame to group by.\n",
      " |      \n",
      " |          .. versionchanged:: 1.4.0\n",
      " |      \n",
      " |             Previously, `by` is silently ignore and makes no groupings\n",
      " |      \n",
      " |      **kwargs\n",
      " |          Additional keywords are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`matplotlib.axes.Axes` or numpy.ndarray of them\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      DataFrame.boxplot: Another method to draw a box plot.\n",
      " |      Series.plot.box: Draw a box plot from a Series object.\n",
      " |      matplotlib.pyplot.boxplot: Draw a box plot in matplotlib.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Draw a box plot from a DataFrame with four columns of randomly\n",
      " |      generated data.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> data = np.random.randn(25, 4)\n",
      " |          >>> df = pd.DataFrame(data, columns=list('ABCD'))\n",
      " |          >>> ax = df.plot.box()\n",
      " |      \n",
      " |      You can also generate groupings if you specify the `by` parameter (which\n",
      " |      can take a column name, or a list or tuple of column names):\n",
      " |      \n",
      " |      .. versionchanged:: 1.4.0\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> age_list = [8, 10, 12, 14, 72, 74, 76, 78, 20, 25, 30, 35, 60, 85]\n",
      " |          >>> df = pd.DataFrame({\"gender\": list(\"MMMMMMMMFFFFFF\"), \"age\": age_list})\n",
      " |          >>> ax = df.plot.box(column=\"age\", by=\"gender\", figsize=(10, 8))\n",
      " |  \n",
      " |  density = kde(self, bw_method=None, ind=None, **kwargs)\n",
      " |  \n",
      " |  hexbin(self, x, y, C=None, reduce_C_function=None, gridsize=None, **kwargs)\n",
      " |      Generate a hexagonal binning plot.\n",
      " |      \n",
      " |      Generate a hexagonal binning plot of `x` versus `y`. If `C` is `None`\n",
      " |      (the default), this is a histogram of the number of occurrences\n",
      " |      of the observations at ``(x[i], y[i])``.\n",
      " |      \n",
      " |      If `C` is specified, specifies values at given coordinates\n",
      " |      ``(x[i], y[i])``. These values are accumulated for each hexagonal\n",
      " |      bin and then reduced according to `reduce_C_function`,\n",
      " |      having as default the NumPy's mean function (:meth:`numpy.mean`).\n",
      " |      (If `C` is specified, it must also be a 1-D sequence\n",
      " |      of the same length as `x` and `y`, or a column label.)\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      x : int or str\n",
      " |          The column label or position for x points.\n",
      " |      y : int or str\n",
      " |          The column label or position for y points.\n",
      " |      C : int or str, optional\n",
      " |          The column label or position for the value of `(x, y)` point.\n",
      " |      reduce_C_function : callable, default `np.mean`\n",
      " |          Function of one argument that reduces all the values in a bin to\n",
      " |          a single number (e.g. `np.mean`, `np.max`, `np.sum`, `np.std`).\n",
      " |      gridsize : int or tuple of (int, int), default 100\n",
      " |          The number of hexagons in the x-direction.\n",
      " |          The corresponding number of hexagons in the y-direction is\n",
      " |          chosen in a way that the hexagons are approximately regular.\n",
      " |          Alternatively, gridsize can be a tuple with two elements\n",
      " |          specifying the number of hexagons in the x-direction and the\n",
      " |          y-direction.\n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.AxesSubplot\n",
      " |          The matplotlib ``Axes`` on which the hexbin is plotted.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      DataFrame.plot : Make plots of a DataFrame.\n",
      " |      matplotlib.pyplot.hexbin : Hexagonal binning plot using matplotlib,\n",
      " |          the matplotlib function that is used under the hood.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      The following examples are generated with random data from\n",
      " |      a normal distribution.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> n = 10000\n",
      " |          >>> df = pd.DataFrame({'x': np.random.randn(n),\n",
      " |          ...                    'y': np.random.randn(n)})\n",
      " |          >>> ax = df.plot.hexbin(x='x', y='y', gridsize=20)\n",
      " |      \n",
      " |      The next example uses `C` and `np.sum` as `reduce_C_function`.\n",
      " |      Note that `'observations'` values ranges from 1 to 5 but the result\n",
      " |      plot shows values up to more than 25. This is because of the\n",
      " |      `reduce_C_function`.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> n = 500\n",
      " |          >>> df = pd.DataFrame({\n",
      " |          ...     'coord_x': np.random.uniform(-3, 3, size=n),\n",
      " |          ...     'coord_y': np.random.uniform(30, 50, size=n),\n",
      " |          ...     'observations': np.random.randint(1,5, size=n)\n",
      " |          ...     })\n",
      " |          >>> ax = df.plot.hexbin(x='coord_x',\n",
      " |          ...                     y='coord_y',\n",
      " |          ...                     C='observations',\n",
      " |          ...                     reduce_C_function=np.sum,\n",
      " |          ...                     gridsize=10,\n",
      " |          ...                     cmap=\"viridis\")\n",
      " |  \n",
      " |  hist(self, by=None, bins=10, **kwargs)\n",
      " |      Draw one histogram of the DataFrame's columns.\n",
      " |      \n",
      " |      A histogram is a representation of the distribution of data.\n",
      " |      This function groups the values of all given Series in the DataFrame\n",
      " |      into bins and draws all bins in one :class:`matplotlib.axes.Axes`.\n",
      " |      This is useful when the DataFrame's Series are in a similar scale.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      by : str or sequence, optional\n",
      " |          Column in the DataFrame to group by.\n",
      " |      \n",
      " |          .. versionchanged:: 1.4.0\n",
      " |      \n",
      " |             Previously, `by` is silently ignore and makes no groupings\n",
      " |      \n",
      " |      bins : int, default 10\n",
      " |          Number of histogram bins to be used.\n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      class:`matplotlib.AxesSubplot`\n",
      " |          Return a histogram plot.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      DataFrame.hist : Draw histograms per DataFrame's Series.\n",
      " |      Series.hist : Draw a histogram with Series' data.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      When we roll a die 6000 times, we expect to get each value around 1000\n",
      " |      times. But when we roll two dice and sum the result, the distribution\n",
      " |      is going to be quite different. A histogram illustrates those\n",
      " |      distributions.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> df = pd.DataFrame(\n",
      " |          ...     np.random.randint(1, 7, 6000),\n",
      " |          ...     columns = ['one'])\n",
      " |          >>> df['two'] = df['one'] + np.random.randint(1, 7, 6000)\n",
      " |          >>> ax = df.plot.hist(bins=12, alpha=0.5)\n",
      " |      \n",
      " |      A grouped histogram can be generated by providing the parameter `by` (which\n",
      " |      can be a column name, or a list of column names):\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> age_list = [8, 10, 12, 14, 72, 74, 76, 78, 20, 25, 30, 35, 60, 85]\n",
      " |          >>> df = pd.DataFrame({\"gender\": list(\"MMMMMMMMFFFFFF\"), \"age\": age_list})\n",
      " |          >>> ax = df.plot.hist(column=[\"age\"], by=\"gender\", figsize=(10, 8))\n",
      " |  \n",
      " |  kde(self, bw_method=None, ind=None, **kwargs)\n",
      " |      Generate Kernel Density Estimate plot using Gaussian kernels.\n",
      " |      \n",
      " |      In statistics, `kernel density estimation`_ (KDE) is a non-parametric\n",
      " |      way to estimate the probability density function (PDF) of a random\n",
      " |      variable. This function uses Gaussian kernels and includes automatic\n",
      " |      bandwidth determination.\n",
      " |      \n",
      " |      .. _kernel density estimation:\n",
      " |          https://en.wikipedia.org/wiki/Kernel_density_estimation\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      bw_method : str, scalar or callable, optional\n",
      " |          The method used to calculate the estimator bandwidth. This can be\n",
      " |          'scott', 'silverman', a scalar constant or a callable.\n",
      " |          If None (default), 'scott' is used.\n",
      " |          See :class:`scipy.stats.gaussian_kde` for more information.\n",
      " |      ind : NumPy array or int, optional\n",
      " |          Evaluation points for the estimated PDF. If None (default),\n",
      " |          1000 equally spaced points are used. If `ind` is a NumPy array, the\n",
      " |          KDE is evaluated at the points passed. If `ind` is an integer,\n",
      " |          `ind` number of equally spaced points are used.\n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.axes.Axes or numpy.ndarray of them\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      scipy.stats.gaussian_kde : Representation of a kernel-density\n",
      " |          estimate using Gaussian kernels. This is the function used\n",
      " |          internally to estimate the PDF.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Given a Series of points randomly sampled from an unknown\n",
      " |      distribution, estimate its PDF using KDE with automatic\n",
      " |      bandwidth determination and plot the results, evaluating them at\n",
      " |      1000 equally spaced points (default):\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> s = pd.Series([1, 2, 2.5, 3, 3.5, 4, 5])\n",
      " |          >>> ax = s.plot.kde()\n",
      " |      \n",
      " |      A scalar bandwidth can be specified. Using a small bandwidth value can\n",
      " |      lead to over-fitting, while using a large bandwidth value may result\n",
      " |      in under-fitting:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = s.plot.kde(bw_method=0.3)\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = s.plot.kde(bw_method=3)\n",
      " |      \n",
      " |      Finally, the `ind` parameter determines the evaluation points for the\n",
      " |      plot of the estimated PDF:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = s.plot.kde(ind=[1, 2, 3, 4, 5])\n",
      " |      \n",
      " |      For DataFrame, it works in the same way:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> df = pd.DataFrame({\n",
      " |          ...     'x': [1, 2, 2.5, 3, 3.5, 4, 5],\n",
      " |          ...     'y': [4, 4, 4.5, 5, 5.5, 6, 6],\n",
      " |          ... })\n",
      " |          >>> ax = df.plot.kde()\n",
      " |      \n",
      " |      A scalar bandwidth can be specified. Using a small bandwidth value can\n",
      " |      lead to over-fitting, while using a large bandwidth value may result\n",
      " |      in under-fitting:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = df.plot.kde(bw_method=0.3)\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = df.plot.kde(bw_method=3)\n",
      " |      \n",
      " |      Finally, the `ind` parameter determines the evaluation points for the\n",
      " |      plot of the estimated PDF:\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax = df.plot.kde(ind=[1, 2, 3, 4, 5, 6])\n",
      " |  \n",
      " |  line(self, x=None, y=None, **kwargs)\n",
      " |      Plot Series or DataFrame as lines.\n",
      " |      \n",
      " |      This function is useful to plot lines using DataFrame's values\n",
      " |      as coordinates.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      x : label or position, optional\n",
      " |          Allows plotting of one column versus another. If not specified,\n",
      " |          the index of the DataFrame is used.\n",
      " |      y : label or position, optional\n",
      " |          Allows plotting of one column versus another. If not specified,\n",
      " |          all numerical columns are used.\n",
      " |      color : str, array-like, or dict, optional\n",
      " |          The color for each of the DataFrame's columns. Possible values are:\n",
      " |      \n",
      " |          - A single color string referred to by name, RGB or RGBA code,\n",
      " |              for instance 'red' or '#a98d19'.\n",
      " |      \n",
      " |          - A sequence of color strings referred to by name, RGB or RGBA\n",
      " |              code, which will be used for each column recursively. For\n",
      " |              instance ['green','yellow'] each column's line will be filled in\n",
      " |              green or yellow, alternatively. If there is only a single column to\n",
      " |              be plotted, then only the first color from the color list will be\n",
      " |              used.\n",
      " |      \n",
      " |          - A dict of the form {column name : color}, so that each column will be\n",
      " |              colored accordingly. For example, if your columns are called `a` and\n",
      " |              `b`, then passing {'a': 'green', 'b': 'red'} will color lines for\n",
      " |              column `a` in green and lines for column `b` in red.\n",
      " |      \n",
      " |          .. versionadded:: 1.1.0\n",
      " |      \n",
      " |      **kwargs\n",
      " |          Additional keyword arguments are documented in\n",
      " |          :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.axes.Axes or np.ndarray of them\n",
      " |          An ndarray is returned with one :class:`matplotlib.axes.Axes`\n",
      " |          per column when ``subplots=True``.\n",
      " |      \n",
      " |              See Also\n",
      " |              --------\n",
      " |              matplotlib.pyplot.plot : Plot y versus x as lines and/or markers.\n",
      " |      \n",
      " |              Examples\n",
      " |              --------\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  >>> s = pd.Series([1, 3, 2])\n",
      " |                  >>> s.plot.line()\n",
      " |                  <AxesSubplot:ylabel='Density'>\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  The following example shows the populations for some animals\n",
      " |                  over the years.\n",
      " |      \n",
      " |                  >>> df = pd.DataFrame({\n",
      " |                  ...    'pig': [20, 18, 489, 675, 1776],\n",
      " |                  ...    'horse': [4, 25, 281, 600, 1900]\n",
      " |                  ...    }, index=[1990, 1997, 2003, 2009, 2014])\n",
      " |                  >>> lines = df.plot.line()\n",
      " |      \n",
      " |              .. plot::\n",
      " |                 :context: close-figs\n",
      " |      \n",
      " |                 An example with subplots, so an array of axes is returned.\n",
      " |      \n",
      " |                 >>> axes = df.plot.line(subplots=True)\n",
      " |                 >>> type(axes)\n",
      " |                 <class 'numpy.ndarray'>\n",
      " |      \n",
      " |              .. plot::\n",
      " |                 :context: close-figs\n",
      " |      \n",
      " |                 Let's repeat the same example, but specifying colors for\n",
      " |                 each column (in this case, for each animal).\n",
      " |      \n",
      " |                 >>> axes = df.plot.line(\n",
      " |                 ...     subplots=True, color={\"pig\": \"pink\", \"horse\": \"#742802\"}\n",
      " |                 ... )\n",
      " |      \n",
      " |              .. plot::\n",
      " |                  :context: close-figs\n",
      " |      \n",
      " |                  The following example shows the relationship between both\n",
      " |                  populations.\n",
      " |      \n",
      " |                  >>> lines = df.plot.line(x='pig', y='horse')\n",
      " |  \n",
      " |  pie(self, **kwargs)\n",
      " |      Generate a pie plot.\n",
      " |      \n",
      " |      A pie plot is a proportional representation of the numerical data in a\n",
      " |      column. This function wraps :meth:`matplotlib.pyplot.pie` for the\n",
      " |      specified column. If no column reference is passed and\n",
      " |      ``subplots=True`` a pie plot is drawn for each numerical column\n",
      " |      independently.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      y : int or label, optional\n",
      " |          Label or position of the column to plot.\n",
      " |          If not provided, ``subplots=True`` argument must be passed.\n",
      " |      **kwargs\n",
      " |          Keyword arguments to pass on to :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      matplotlib.axes.Axes or np.ndarray of them\n",
      " |          A NumPy array is returned when `subplots` is True.\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      Series.plot.pie : Generate a pie plot for a Series.\n",
      " |      DataFrame.plot : Make plots of a DataFrame.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      In the example below we have a DataFrame with the information about\n",
      " |      planet's mass and radius. We pass the 'mass' column to the\n",
      " |      pie function to get a pie plot.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> df = pd.DataFrame({'mass': [0.330, 4.87 , 5.97],\n",
      " |          ...                    'radius': [2439.7, 6051.8, 6378.1]},\n",
      " |          ...                   index=['Mercury', 'Venus', 'Earth'])\n",
      " |          >>> plot = df.plot.pie(y='mass', figsize=(5, 5))\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> plot = df.plot.pie(subplots=True, figsize=(11, 6))\n",
      " |  \n",
      " |  scatter(self, x, y, s=None, c=None, **kwargs)\n",
      " |      Create a scatter plot with varying marker point size and color.\n",
      " |      \n",
      " |      The coordinates of each point are defined by two dataframe columns and\n",
      " |      filled circles are used to represent each point. This kind of plot is\n",
      " |      useful to see complex correlations between two variables. Points could\n",
      " |      be for instance natural 2D coordinates like longitude and latitude in\n",
      " |      a map or, in general, any pair of metrics that can be plotted against\n",
      " |      each other.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      x : int or str\n",
      " |          The column name or column position to be used as horizontal\n",
      " |          coordinates for each point.\n",
      " |      y : int or str\n",
      " |          The column name or column position to be used as vertical\n",
      " |          coordinates for each point.\n",
      " |      s : str, scalar or array-like, optional\n",
      " |          The size of each point. Possible values are:\n",
      " |      \n",
      " |          - A string with the name of the column to be used for marker's size.\n",
      " |      \n",
      " |          - A single scalar so all points have the same size.\n",
      " |      \n",
      " |          - A sequence of scalars, which will be used for each point's size\n",
      " |            recursively. For instance, when passing [2,14] all points size\n",
      " |            will be either 2 or 14, alternatively.\n",
      " |      \n",
      " |            .. versionchanged:: 1.1.0\n",
      " |      \n",
      " |      c : str, int or array-like, optional\n",
      " |          The color of each point. Possible values are:\n",
      " |      \n",
      " |          - A single color string referred to by name, RGB or RGBA code,\n",
      " |            for instance 'red' or '#a98d19'.\n",
      " |      \n",
      " |          - A sequence of color strings referred to by name, RGB or RGBA\n",
      " |            code, which will be used for each point's color recursively. For\n",
      " |            instance ['green','yellow'] all points will be filled in green or\n",
      " |            yellow, alternatively.\n",
      " |      \n",
      " |          - A column name or position whose values will be used to color the\n",
      " |            marker points according to a colormap.\n",
      " |      \n",
      " |      **kwargs\n",
      " |          Keyword arguments to pass on to :meth:`DataFrame.plot`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`matplotlib.axes.Axes` or numpy.ndarray of them\n",
      " |      \n",
      " |      See Also\n",
      " |      --------\n",
      " |      matplotlib.pyplot.scatter : Scatter plot using multiple input data\n",
      " |          formats.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      Let's see how to draw a scatter plot using coordinates from the values\n",
      " |      in a DataFrame's columns.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> df = pd.DataFrame([[5.1, 3.5, 0], [4.9, 3.0, 0], [7.0, 3.2, 1],\n",
      " |          ...                    [6.4, 3.2, 1], [5.9, 3.0, 2]],\n",
      " |          ...                   columns=['length', 'width', 'species'])\n",
      " |          >>> ax1 = df.plot.scatter(x='length',\n",
      " |          ...                       y='width',\n",
      " |          ...                       c='DarkBlue')\n",
      " |      \n",
      " |      And now with the color determined by a column as well.\n",
      " |      \n",
      " |      .. plot::\n",
      " |          :context: close-figs\n",
      " |      \n",
      " |          >>> ax2 = df.plot.scatter(x='length',\n",
      " |          ...                       y='width',\n",
      " |          ...                       c='species',\n",
      " |          ...                       colormap='viridis')\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __annotations__ = {}\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pandas.core.base.PandasObject:\n",
      " |  \n",
      " |  __repr__(self) -> 'str'\n",
      " |      Return a string representation for a particular object.\n",
      " |  \n",
      " |  __sizeof__(self) -> 'int'\n",
      " |      Generates the total memory usage for an object that returns\n",
      " |      either a value or Series of values\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from pandas.core.accessor.DirNamesMixin:\n",
      " |  \n",
      " |  __dir__(self) -> 'list[str]'\n",
      " |      Provide method name lookup and completion.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Only provide 'public' methods.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from pandas.core.accessor.DirNamesMixin:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(df.plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a\n",
       "a  0\n",
       "b  9"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 0 ns, sys: 1 µs, total: 1 µs\n",
      "Wall time: 2.15 µs\n",
      "28\n",
      "553942\n",
      "387157\n",
      "506655\n",
      "522019\n",
      "436605\n",
      "503589\n",
      "418377\n",
      "570555\n",
      "641810\n",
      "633009\n",
      "498685\n",
      "530719\n",
      "468912\n",
      "627153\n",
      "487607\n",
      "607455\n",
      "632898\n",
      "614981\n",
      "432772\n",
      "489456\n",
      "572237\n",
      "716704\n",
      "508945\n",
      "378263\n",
      "410248\n",
      "587297\n",
      "510855\n",
      "597782\n"
     ]
    }
   ],
   "source": [
    "%time\n",
    "sample_id = 'sample_id'\n",
    "p='/Users/liam/Downloads/functional/'\n",
    "gzs =glob(p+'BBB*.MICROBA.sample.tsv.gz')\n",
    "print(len(gzs))\n",
    "\n",
    "for i in gzs:\n",
    "\n",
    "    df = pd.read_csv(i, index_col =0, sep='\\t')\n",
    "    assert len(df.columns) == 1\n",
    "    df.columns = [sample_id]\n",
    "\n",
    "    print(len(df[sample_id]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIG20000000028</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20000000056</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20000000075</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20000000096</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20000000124</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20073909726</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20073909781</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20073909783</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20073909786</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIG20073911408</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>489456 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Count\n",
       "ID                   \n",
       "MIG20000000028      1\n",
       "MIG20000000056      4\n",
       "MIG20000000075      1\n",
       "MIG20000000096      2\n",
       "MIG20000000124      3\n",
       "...               ...\n",
       "MIG20073909726      1\n",
       "MIG20073909781     10\n",
       "MIG20073909783      1\n",
       "MIG20073909786      5\n",
       "MIG20073911408     12\n",
       "\n",
       "[489456 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/liam/Downloads/BBB6542-PBC4-LBC2-R006-FSS.MICROBA.sample.tsv.gz', index_col =0, sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 fu\n"
     ]
    }
   ],
   "source": [
    "@dataclass\n",
    "class C2:\n",
    "    x: int\n",
    "    def f(self, y):\n",
    "        print(self.x*y)\n",
    "\n",
    "@dataclass\n",
    "class C1(C2):\n",
    "    def f(self, y):\n",
    "        print(self.x+y, 'fu')\n",
    "\n",
    "c1 = C1(1)\n",
    "c1.f(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "d={'a': C2}\n",
    "c = d.get('a', C1)(1)\n",
    "c.f(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "CPY_BACKGROUND_SPECIES = \"Taxon\\n__Pseudomonas aeruginosa\\ns__Escherichia coli\\ns__Salmonella enterica\\ns__Lactobacillus_H fermentum\\ns__Enterococcus faecalis\\ns__Staphylococcus aureus\\ns__Listeria monocytogenes_A\\ns__Listeria monocytogenes_B\\ns__Listeria monocytogenes\\ns__Bacillus subtilis_D\\ns__Saccharomyces group A\\ns__Cryptococcus neoformans\\ns__Bacillus marinus\\ns__Escherichia flexneri\"\n",
    "CPD_BACKGROUND_SPECIES = \"Taxon\\ns__Acinetobacter baumannii\\ns__Actinomyces_B odontolyticus\\ns__Bacillus_A cereus_AD\\ns__Bacteroides_B vulgatus\\ns__Bifidobacterium adolescentis\\ns__Clostridium beijerinckii\\ns__Cutibacterium acnes\\ns__Deinococcus radiodurans\\ns__Enterococcus faecalis\\ns__Escherichia coli\\ns__Helicobacter pylori_AS\\ns__Lactobacillus gasseri\\ns__Neisseria meningitidis\\ns__Porphyromonas gingivalis\\ns__Pseudomonas aeruginosa_A\\ns__Staphylococcus aureus\\ns__Staphylococcus epidermidis\\ns__Streptococcus agalactiae\\ns__Streptococcus mutans\\ns__Helicobacter pylori\\ns__Escherichia flexneri\\ns__Pauljensenia odontolyticus_A\"\n",
    "CPD_BACKGROUND_SPECIES2 = \"Taxon\\ns__Acinetobacter ff\\ns__Actinomyces_B ff\\ns__Bacillus_A ff\\ns__Bacteroides_B vulgatus\\ns__Bifidobacterium adolescentis\\ns__Clostridium beijerinckii\\ns__Cutibacterium acnes\\ns__Deinococcus radiodurans\\ns__Enterococcus faecalis\\ns__Escherichia coli\\ns__Helicobacter pylori_AS\\ns__Lactobacillus gasseri\\ns__Neisseria meningitidis\\ns__Porphyromonas gingivalis\\ns__Pseudomonas aeruginosa_A\\ns__Staphylococcus aureus\\ns__Staphylococcus epidermidis\\ns__Streptococcus agalactiae\\ns__Streptococcus mutans\\ns__Helicobacter pylori\\ns__Escherichia flexneri\\ns__Pauljensenia odontolyticus_A\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>b2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Taxon</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Taxon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>__Pseudomonas aeruginosa</td>\n",
       "      <td>s__Acinetobacter baumannii</td>\n",
       "      <td>s__Acinetobacter ff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s__Escherichia coli</td>\n",
       "      <td>s__Actinomyces_B odontolyticus</td>\n",
       "      <td>s__Actinomyces_B ff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s__Salmonella enterica</td>\n",
       "      <td>s__Bacillus_A cereus_AD</td>\n",
       "      <td>s__Bacillus_A ff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s__Lactobacillus_H fermentum</td>\n",
       "      <td>s__Bacteroides_B vulgatus</td>\n",
       "      <td>s__Bacteroides_B vulgatus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              a                               b  \\\n",
       "0                         Taxon                           Taxon   \n",
       "1      __Pseudomonas aeruginosa      s__Acinetobacter baumannii   \n",
       "2           s__Escherichia coli  s__Actinomyces_B odontolyticus   \n",
       "3        s__Salmonella enterica         s__Bacillus_A cereus_AD   \n",
       "4  s__Lactobacillus_H fermentum       s__Bacteroides_B vulgatus   \n",
       "\n",
       "                          b2  \n",
       "0                      Taxon  \n",
       "1        s__Acinetobacter ff  \n",
       "2        s__Actinomyces_B ff  \n",
       "3           s__Bacillus_A ff  \n",
       "4  s__Bacteroides_B vulgatus  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df =pd.DataFrame([CPY_BACKGROUND_SPECIES.split('\\n'), CPD_BACKGROUND_SPECIES.split('\\n'), CPD_BACKGROUND_SPECIES2.split('\\n')]).T\n",
    "df.columns = ['a', 'b', 'b2']\n",
    "df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5k/61vjt_z14fg3thmw_wts19f40000gp/T/ipykernel_41367/2934274232.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.\n",
      "  df[\"exists\"] = df.drop(\"b2\", 1).isin(df[\"b2\"]).any(1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>b2</th>\n",
       "      <th>exists</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Taxon</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>__Pseudomonas aeruginosa</td>\n",
       "      <td>s__Acinetobacter baumannii</td>\n",
       "      <td>s__Acinetobacter ff</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s__Escherichia coli</td>\n",
       "      <td>s__Actinomyces_B odontolyticus</td>\n",
       "      <td>s__Actinomyces_B ff</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s__Salmonella enterica</td>\n",
       "      <td>s__Bacillus_A cereus_AD</td>\n",
       "      <td>s__Bacillus_A ff</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>s__Lactobacillus_H fermentum</td>\n",
       "      <td>s__Bacteroides_B vulgatus</td>\n",
       "      <td>s__Bacteroides_B vulgatus</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>s__Enterococcus faecalis</td>\n",
       "      <td>s__Bifidobacterium adolescentis</td>\n",
       "      <td>s__Bifidobacterium adolescentis</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>s__Staphylococcus aureus</td>\n",
       "      <td>s__Clostridium beijerinckii</td>\n",
       "      <td>s__Clostridium beijerinckii</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>s__Listeria monocytogenes_A</td>\n",
       "      <td>s__Cutibacterium acnes</td>\n",
       "      <td>s__Cutibacterium acnes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>s__Listeria monocytogenes_B</td>\n",
       "      <td>s__Deinococcus radiodurans</td>\n",
       "      <td>s__Deinococcus radiodurans</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>s__Listeria monocytogenes</td>\n",
       "      <td>s__Enterococcus faecalis</td>\n",
       "      <td>s__Enterococcus faecalis</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>s__Bacillus subtilis_D</td>\n",
       "      <td>s__Escherichia coli</td>\n",
       "      <td>s__Escherichia coli</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>s__Saccharomyces group A</td>\n",
       "      <td>s__Helicobacter pylori_AS</td>\n",
       "      <td>s__Helicobacter pylori_AS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>s__Cryptococcus neoformans</td>\n",
       "      <td>s__Lactobacillus gasseri</td>\n",
       "      <td>s__Lactobacillus gasseri</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>s__Bacillus marinus</td>\n",
       "      <td>s__Neisseria meningitidis</td>\n",
       "      <td>s__Neisseria meningitidis</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>s__Escherichia flexneri</td>\n",
       "      <td>s__Porphyromonas gingivalis</td>\n",
       "      <td>s__Porphyromonas gingivalis</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Pseudomonas aeruginosa_A</td>\n",
       "      <td>s__Pseudomonas aeruginosa_A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Staphylococcus aureus</td>\n",
       "      <td>s__Staphylococcus aureus</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Staphylococcus epidermidis</td>\n",
       "      <td>s__Staphylococcus epidermidis</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Streptococcus agalactiae</td>\n",
       "      <td>s__Streptococcus agalactiae</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Streptococcus mutans</td>\n",
       "      <td>s__Streptococcus mutans</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Helicobacter pylori</td>\n",
       "      <td>s__Helicobacter pylori</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Escherichia flexneri</td>\n",
       "      <td>s__Escherichia flexneri</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>None</td>\n",
       "      <td>s__Pauljensenia odontolyticus_A</td>\n",
       "      <td>s__Pauljensenia odontolyticus_A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               a                                b  \\\n",
       "0                          Taxon                            Taxon   \n",
       "1       __Pseudomonas aeruginosa       s__Acinetobacter baumannii   \n",
       "2            s__Escherichia coli   s__Actinomyces_B odontolyticus   \n",
       "3         s__Salmonella enterica          s__Bacillus_A cereus_AD   \n",
       "4   s__Lactobacillus_H fermentum        s__Bacteroides_B vulgatus   \n",
       "5       s__Enterococcus faecalis  s__Bifidobacterium adolescentis   \n",
       "6       s__Staphylococcus aureus      s__Clostridium beijerinckii   \n",
       "7    s__Listeria monocytogenes_A           s__Cutibacterium acnes   \n",
       "8    s__Listeria monocytogenes_B       s__Deinococcus radiodurans   \n",
       "9      s__Listeria monocytogenes         s__Enterococcus faecalis   \n",
       "10        s__Bacillus subtilis_D              s__Escherichia coli   \n",
       "11      s__Saccharomyces group A        s__Helicobacter pylori_AS   \n",
       "12    s__Cryptococcus neoformans         s__Lactobacillus gasseri   \n",
       "13           s__Bacillus marinus        s__Neisseria meningitidis   \n",
       "14       s__Escherichia flexneri      s__Porphyromonas gingivalis   \n",
       "15                          None      s__Pseudomonas aeruginosa_A   \n",
       "16                          None         s__Staphylococcus aureus   \n",
       "17                          None    s__Staphylococcus epidermidis   \n",
       "18                          None      s__Streptococcus agalactiae   \n",
       "19                          None          s__Streptococcus mutans   \n",
       "20                          None           s__Helicobacter pylori   \n",
       "21                          None          s__Escherichia flexneri   \n",
       "22                          None  s__Pauljensenia odontolyticus_A   \n",
       "\n",
       "                                 b2  exists  \n",
       "0                             Taxon    True  \n",
       "1               s__Acinetobacter ff   False  \n",
       "2               s__Actinomyces_B ff   False  \n",
       "3                  s__Bacillus_A ff   False  \n",
       "4         s__Bacteroides_B vulgatus    True  \n",
       "5   s__Bifidobacterium adolescentis    True  \n",
       "6       s__Clostridium beijerinckii    True  \n",
       "7            s__Cutibacterium acnes    True  \n",
       "8        s__Deinococcus radiodurans    True  \n",
       "9          s__Enterococcus faecalis    True  \n",
       "10              s__Escherichia coli    True  \n",
       "11        s__Helicobacter pylori_AS    True  \n",
       "12         s__Lactobacillus gasseri    True  \n",
       "13        s__Neisseria meningitidis    True  \n",
       "14      s__Porphyromonas gingivalis    True  \n",
       "15      s__Pseudomonas aeruginosa_A    True  \n",
       "16         s__Staphylococcus aureus    True  \n",
       "17    s__Staphylococcus epidermidis    True  \n",
       "18      s__Streptococcus agalactiae    True  \n",
       "19          s__Streptococcus mutans    True  \n",
       "20           s__Helicobacter pylori    True  \n",
       "21          s__Escherichia flexneri    True  \n",
       "22  s__Pauljensenia odontolyticus_A    True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"exists\"] = df.drop(\"b2\", 1).isin(df[\"b2\"]).any(1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/liam/Downloads/all_trusted_flex_v2_r322.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/liam/code/Untitled-1.ipynb Cell 7\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/liam/code/Untitled-1.ipynb#ch0000006?line=0'>1</a>\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_csv(\u001b[39m'\u001b[39;49m\u001b[39m/Users/liam/Downloads/all_trusted_flex_v2_r322.txt\u001b[39;49m\u001b[39m'\u001b[39;49m, header\u001b[39m=\u001b[39;49m\u001b[39mNone\u001b[39;49;00m, names\u001b[39m=\u001b[39;49m[\u001b[39m'\u001b[39;49m\u001b[39mcontrols\u001b[39;49m\u001b[39m'\u001b[39;49m])\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/liam/code/Untitled-1.ipynb#ch0000006?line=1'>2</a>\u001b[0m df[[\u001b[39m'\u001b[39m\u001b[39mname\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mfoo\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mbar\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mrun\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mCTL\u001b[39m\u001b[39m'\u001b[39m]] \u001b[39m=\u001b[39m df\u001b[39m.\u001b[39mcontrols\u001b[39m.\u001b[39mstr\u001b[39m.\u001b[39msplit(\u001b[39m'\u001b[39m\u001b[39m-\u001b[39m\u001b[39m'\u001b[39m, expand\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/liam/code/Untitled-1.ipynb#ch0000006?line=2'>3</a>\u001b[0m df\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(args) \u001b[39m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[39m.\u001b[39mformat(arguments\u001b[39m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[39mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[39m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:680\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    665\u001b[0m kwds_defaults \u001b[39m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    666\u001b[0m     dialect,\n\u001b[1;32m    667\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    676\u001b[0m     defaults\u001b[39m=\u001b[39m{\u001b[39m\"\u001b[39m\u001b[39mdelimiter\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m\"\u001b[39m\u001b[39m,\u001b[39m\u001b[39m\"\u001b[39m},\n\u001b[1;32m    677\u001b[0m )\n\u001b[1;32m    678\u001b[0m kwds\u001b[39m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 680\u001b[0m \u001b[39mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:575\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    572\u001b[0m _validate_names(kwds\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mnames\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m))\n\u001b[1;32m    574\u001b[0m \u001b[39m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 575\u001b[0m parser \u001b[39m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwds)\n\u001b[1;32m    577\u001b[0m \u001b[39mif\u001b[39;00m chunksize \u001b[39mor\u001b[39;00m iterator:\n\u001b[1;32m    578\u001b[0m     \u001b[39mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:933\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    930\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39moptions[\u001b[39m\"\u001b[39m\u001b[39mhas_index_names\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m kwds[\u001b[39m\"\u001b[39m\u001b[39mhas_index_names\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[1;32m    932\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles: IOHandles \u001b[39m|\u001b[39m \u001b[39mNone\u001b[39;00m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m--> 933\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_engine(f, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mengine)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1217\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1213\u001b[0m     mode \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1214\u001b[0m \u001b[39m# error: No overload variant of \"get_handle\" matches argument types\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m \u001b[39m# \"Union[str, PathLike[str], ReadCsvBuffer[bytes], ReadCsvBuffer[str]]\"\u001b[39;00m\n\u001b[1;32m   1216\u001b[0m \u001b[39m# , \"str\", \"bool\", \"Any\", \"Any\", \"Any\", \"Any\", \"Any\"\u001b[39;00m\n\u001b[0;32m-> 1217\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39m=\u001b[39m get_handle(  \u001b[39m# type: ignore[call-overload]\u001b[39;49;00m\n\u001b[1;32m   1218\u001b[0m     f,\n\u001b[1;32m   1219\u001b[0m     mode,\n\u001b[1;32m   1220\u001b[0m     encoding\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1221\u001b[0m     compression\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mcompression\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1222\u001b[0m     memory_map\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mmemory_map\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mFalse\u001b[39;49;00m),\n\u001b[1;32m   1223\u001b[0m     is_text\u001b[39m=\u001b[39;49mis_text,\n\u001b[1;32m   1224\u001b[0m     errors\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding_errors\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mstrict\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[1;32m   1225\u001b[0m     storage_options\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mstorage_options\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[1;32m   1226\u001b[0m )\n\u001b[1;32m   1227\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m   1228\u001b[0m f \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles\u001b[39m.\u001b[39mhandle\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pandas/io/common.py:789\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    784\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(handle, \u001b[39mstr\u001b[39m):\n\u001b[1;32m    785\u001b[0m     \u001b[39m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    786\u001b[0m     \u001b[39m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    787\u001b[0m     \u001b[39mif\u001b[39;00m ioargs\u001b[39m.\u001b[39mencoding \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m ioargs\u001b[39m.\u001b[39mmode:\n\u001b[1;32m    788\u001b[0m         \u001b[39m# Encoding\u001b[39;00m\n\u001b[0;32m--> 789\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(\n\u001b[1;32m    790\u001b[0m             handle,\n\u001b[1;32m    791\u001b[0m             ioargs\u001b[39m.\u001b[39;49mmode,\n\u001b[1;32m    792\u001b[0m             encoding\u001b[39m=\u001b[39;49mioargs\u001b[39m.\u001b[39;49mencoding,\n\u001b[1;32m    793\u001b[0m             errors\u001b[39m=\u001b[39;49merrors,\n\u001b[1;32m    794\u001b[0m             newline\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    795\u001b[0m         )\n\u001b[1;32m    796\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    797\u001b[0m         \u001b[39m# Binary mode\u001b[39;00m\n\u001b[1;32m    798\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39m(handle, ioargs\u001b[39m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/liam/Downloads/all_trusted_flex_v2_r322.txt'"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/liam/Downloads/all_trusted_flex_v2_r322.txt', header=None, names=['controls'])\n",
    "df[['name', 'foo', 'bar', 'run', 'CTL']] = df.controls.str.split('-', expand=True)\n",
    "df\n",
    "CTLs=set(df.CTL)\n",
    "if len(CTLs) != 1 and CTLs.pop() != 'CTL':\n",
    "    raise ValueError(f'Type should be CTL, only; {CTLs}')\n",
    "\n",
    "df['control_type'] = df.name.str[:3]\n",
    "df=df[['controls', 'name', 'run', 'control_type']]\n",
    "df.value_counts('control_type')\n",
    "df.sort_values('run', inplace=True)\n",
    "df.tail(50).controls.to_csv('~/Downloads/hhhh.lof', index=False, header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0            s__Acetatifactor sp900066565\n",
       "1          s__Adlercreutzia equolifaciens\n",
       "2                 s__Agathobacter rectale\n",
       "3      s__Agathobaculum butyriciproducens\n",
       "4            s__Akkermansia muciniphila_B\n",
       "                      ...                \n",
       "112                s__UBA1417 sp003531055\n",
       "113                    s__UBA3818 MIC8425\n",
       "114                    s__UBA7160 MIC6745\n",
       "115                    s__UBA7160 MIC9207\n",
       "116                    s__UBA7182 MIC8422\n",
       "Name: Taxon, Length: 117, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/liam/Downloads/R388/CPG/sp_test/profiles_counts_background/profile_summary_stats.tsv', sep='\\t')\n",
    "df.Taxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(369, 42)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>bio_material</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>Submitter_Id</th>\n",
       "      <th>diet</th>\n",
       "      <th>days_post_colonization</th>\n",
       "      <th>donor_microbiota</th>\n",
       "      <th>experimental_group</th>\n",
       "      <th>Isolator</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Collection_day</th>\n",
       "      <th>Participant_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR5261198</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>269</td>\n",
       "      <td>7990595975</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772329</td>\n",
       "      <td>2814020302</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>ERP121534</td>\n",
       "      <td>PF_S.13_Day19</td>\n",
       "      <td>Pea fiber (2 snacks a day)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>S.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5261199</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>265</td>\n",
       "      <td>6859111259</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772328</td>\n",
       "      <td>2276073800</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>ERP121534</td>\n",
       "      <td>PF_S.13_Day18</td>\n",
       "      <td>Pea fiber (2 snacks a day)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>S.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Assay Type  AvgSpotLen       Bases bio_material  BioProject  \\\n",
       "Run                                                                      \n",
       "ERR5261198      OTHER         269  7990595975        Stool  PRJEB38148   \n",
       "ERR5261199      OTHER         265  6859111259        Stool  PRJEB38148   \n",
       "\n",
       "               BioSample       Bytes  \\\n",
       "Run                                    \n",
       "ERR5261198  SAMEA7772329  2814020302   \n",
       "ERR5261199  SAMEA7772328  2276073800   \n",
       "\n",
       "                                                 Center Name Consent  \\\n",
       "Run                                                                    \n",
       "ERR5261198  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "ERR5261199  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "\n",
       "           DATASTORE filetype  ...  SRA Study   Submitter_Id  \\\n",
       "Run                            ...                             \n",
       "ERR5261198          fastq,sra  ...  ERP121534  PF_S.13_Day19   \n",
       "ERR5261199          fastq,sra  ...  ERP121534  PF_S.13_Day18   \n",
       "\n",
       "                                  diet days_post_colonization  \\\n",
       "Run                                                             \n",
       "ERR5261198  Pea fiber (2 snacks a day)                    NaN   \n",
       "ERR5261199  Pea fiber (2 snacks a day)                    NaN   \n",
       "\n",
       "           donor_microbiota experimental_group Isolator Mouse_ID  \\\n",
       "Run                                                                \n",
       "ERR5261198              NaN                NaN      NaN      NaN   \n",
       "ERR5261199              NaN                NaN      NaN      NaN   \n",
       "\n",
       "           Collection_day Participant_ID  \n",
       "Run                                       \n",
       "ERR5261198           19.0           S.13  \n",
       "ERR5261199           18.0           S.13  \n",
       "\n",
       "[2 rows x 42 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfh= pd.read_csv('/Users/liam/Downloads/PRJEB38148_human.csv', index_col=0)\n",
    "print(dfh.shape)\n",
    "\n",
    "dfh.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(363, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR5349076</th>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772053</td>\n",
       "      <td>ERX5137452</td>\n",
       "      <td>HS2</td>\n",
       "      <td>2Fib4Fib_S.09</td>\n",
       "      <td>Day11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5349077</th>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772052</td>\n",
       "      <td>ERX5137453</td>\n",
       "      <td>HS2</td>\n",
       "      <td>2Fib4Fib_S.09</td>\n",
       "      <td>Day10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0             1           2    4              5      6\n",
       "3                                                                          \n",
       "ERR5349076  PRJEB38148  SAMEA7772053  ERX5137452  HS2  2Fib4Fib_S.09  Day11\n",
       "ERR5349077  PRJEB38148  SAMEA7772052  ERX5137453  HS2  2Fib4Fib_S.09  Day10"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfjoel= pd.read_csv('/Users/liam/Downloads/PRJEB38148.metadata.tsv', sep='\\t', index_col=3, header=None)\n",
    "print(dfjoel.shape)\n",
    "\n",
    "dfjoel.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(363, 48)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>bio_material</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>Isolator</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Collection_day</th>\n",
       "      <th>Participant_ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR5261198</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>269</td>\n",
       "      <td>7990595975</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772329</td>\n",
       "      <td>2814020302</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>S.13</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772329</td>\n",
       "      <td>ERX5051565</td>\n",
       "      <td>HS1</td>\n",
       "      <td>PF_S.13</td>\n",
       "      <td>Day19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5261199</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>265</td>\n",
       "      <td>6859111259</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772328</td>\n",
       "      <td>2276073800</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "      <td>S.13</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772328</td>\n",
       "      <td>ERX5051566</td>\n",
       "      <td>HS1</td>\n",
       "      <td>PF_S.13</td>\n",
       "      <td>Day18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5261201</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>264</td>\n",
       "      <td>6016100973</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772330</td>\n",
       "      <td>2001778622</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>S.13</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772330</td>\n",
       "      <td>ERX5051568</td>\n",
       "      <td>HS1</td>\n",
       "      <td>PF_S.13</td>\n",
       "      <td>Day20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5261202</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>264</td>\n",
       "      <td>5721647335</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772334</td>\n",
       "      <td>1886994029</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36.0</td>\n",
       "      <td>S.13</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772334</td>\n",
       "      <td>ERX5051569</td>\n",
       "      <td>HS1</td>\n",
       "      <td>PF_S.13</td>\n",
       "      <td>Day36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR5261204</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>268</td>\n",
       "      <td>6275097722</td>\n",
       "      <td>Stool</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772335</td>\n",
       "      <td>2100540392</td>\n",
       "      <td>WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>S.13</td>\n",
       "      <td>PRJEB38148</td>\n",
       "      <td>SAMEA7772335</td>\n",
       "      <td>ERX5051571</td>\n",
       "      <td>HS1</td>\n",
       "      <td>PF_S.13</td>\n",
       "      <td>Day39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Assay Type  AvgSpotLen       Bases bio_material  BioProject  \\\n",
       "ERR5261198      OTHER         269  7990595975        Stool  PRJEB38148   \n",
       "ERR5261199      OTHER         265  6859111259        Stool  PRJEB38148   \n",
       "ERR5261201      OTHER         264  6016100973        Stool  PRJEB38148   \n",
       "ERR5261202      OTHER         264  5721647335        Stool  PRJEB38148   \n",
       "ERR5261204      OTHER         268  6275097722        Stool  PRJEB38148   \n",
       "\n",
       "               BioSample       Bytes  \\\n",
       "ERR5261198  SAMEA7772329  2814020302   \n",
       "ERR5261199  SAMEA7772328  2276073800   \n",
       "ERR5261201  SAMEA7772330  2001778622   \n",
       "ERR5261202  SAMEA7772334  1886994029   \n",
       "ERR5261204  SAMEA7772335  2100540392   \n",
       "\n",
       "                                                 Center Name Consent  \\\n",
       "ERR5261198  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "ERR5261199  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "ERR5261201  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "ERR5261202  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "ERR5261204  WASHINGTON UNIVERSITY CENTER FOR GENOME SCIENCES  public   \n",
       "\n",
       "           DATASTORE filetype  ... Isolator Mouse_ID Collection_day  \\\n",
       "ERR5261198          fastq,sra  ...      NaN      NaN           19.0   \n",
       "ERR5261199          fastq,sra  ...      NaN      NaN           18.0   \n",
       "ERR5261201          sra,fastq  ...      NaN      NaN           20.0   \n",
       "ERR5261202          sra,fastq  ...      NaN      NaN           36.0   \n",
       "ERR5261204          fastq,sra  ...      NaN      NaN           39.0   \n",
       "\n",
       "           Participant_ID           0             1           2    4        5  \\\n",
       "ERR5261198           S.13  PRJEB38148  SAMEA7772329  ERX5051565  HS1  PF_S.13   \n",
       "ERR5261199           S.13  PRJEB38148  SAMEA7772328  ERX5051566  HS1  PF_S.13   \n",
       "ERR5261201           S.13  PRJEB38148  SAMEA7772330  ERX5051568  HS1  PF_S.13   \n",
       "ERR5261202           S.13  PRJEB38148  SAMEA7772334  ERX5051569  HS1  PF_S.13   \n",
       "ERR5261204           S.13  PRJEB38148  SAMEA7772335  ERX5051571  HS1  PF_S.13   \n",
       "\n",
       "                6  \n",
       "ERR5261198  Day19  \n",
       "ERR5261199  Day18  \n",
       "ERR5261201  Day20  \n",
       "ERR5261202  Day36  \n",
       "ERR5261204  Day39  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dfh.join(dfjoel, how='inner')\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(list())\n",
    "for sample_id, mac_summary in classify_files.items():\n",
    "    try:\n",
    "        summ_df = pd.read_csv(mac_summary, sep=\"\\t\")\n",
    "        if 'None' in summ_df.columns:\n",
    "            self.log.warn(f'No MGDB-ANI-CLASSIFY summary found for sample {sample_id}. Marking as None.')\n",
    "            df.loc[sample_id, 'Query species assignment'] = 'None'\n",
    "            df.loc[sample_id, 'Query taxonomy'] = 'None'\n",
    "            df.loc[sample_id, 'Closest representative'] = 'None'\n",
    "            df.loc[sample_id, 'ANI'] = 'None'\n",
    "            df.loc[sample_id, 'AF'] = 'None'\n",
    "            df.loc[sample_id, 'Representative taxonomy'] = 'None'\n",
    "\n",
    "        else:\n",
    "            df.loc[sample_id, 'Query species assignment'] = summ_df.iloc[0]['Query species assignment']\n",
    "            df.loc[sample_id, 'Query taxonomy'] = summ_df.iloc[0]['Query taxonomy']\n",
    "            df.loc[sample_id, 'Closest representative'] = summ_df.iloc[0]['Closest representative']\n",
    "            df.loc[sample_id, 'ANI'] = summ_df.iloc[0]['ANI']\n",
    "            df.loc[sample_id, 'AF'] = summ_df.iloc[0]['AF']\n",
    "            df.loc[sample_id, 'Representative taxonomy'] = \\\n",
    "                summ_df.iloc[0]['Representative taxonomy']\n",
    "    except Exception:\n",
    "        self.log.error(f\"MGDB-ANI-CLASSIFY results are incomplete. Review file {mac_summary}\")\n",
    "        raise\n",
    "df.index.name = 'Sample'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query species assignment</th>\n",
       "      <th>Query taxonomy</th>\n",
       "      <th>Closest representative</th>\n",
       "      <th>ANI</th>\n",
       "      <th>AF</th>\n",
       "      <th>Representative taxonomy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Query ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BBL6672-PFG3-LFG2-R385-BUG</th>\n",
       "      <td>s__Bifidobacterium adolescentis</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Actinobacte...</td>\n",
       "      <td>G000010425</td>\n",
       "      <td>97.8854</td>\n",
       "      <td>0.857759</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Actinobacte...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Query species assignment  \\\n",
       "Query ID                                                      \n",
       "BBL6672-PFG3-LFG2-R385-BUG  s__Bifidobacterium adolescentis   \n",
       "\n",
       "                                                               Query taxonomy  \\\n",
       "Query ID                                                                        \n",
       "BBL6672-PFG3-LFG2-R385-BUG  d__Bacteria;p__Actinobacteriota;c__Actinobacte...   \n",
       "\n",
       "                           Closest representative      ANI        AF  \\\n",
       "Query ID                                                               \n",
       "BBL6672-PFG3-LFG2-R385-BUG             G000010425  97.8854  0.857759   \n",
       "\n",
       "                                                      Representative taxonomy  \n",
       "Query ID                                                                       \n",
       "BBL6672-PFG3-LFG2-R385-BUG  d__Bacteria;p__Actinobacteriota;c__Actinobacte...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "series = []\n",
    "\n",
    "summ_df = pd.read_csv('/Users/liam/Downloads/testa/BBL6672-PFG3-LFG2-R385-BUG.sp_classification.tsv', sep=\"\\t\", index_col=0)\n",
    "summ_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query species assignment</th>\n",
       "      <th>Query taxonomy</th>\n",
       "      <th>Closest representative</th>\n",
       "      <th>ANI</th>\n",
       "      <th>AF</th>\n",
       "      <th>Representative taxonomy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BBL6672-PFG3-LFG2-R385-BUG</th>\n",
       "      <td>s__Bifidobacterium adolescentis</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Actinobacte...</td>\n",
       "      <td>G000010425</td>\n",
       "      <td>97.8854</td>\n",
       "      <td>0.857759</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Actinobacte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BBL6672-PFG3-LFG2-R385-BUG2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Query species assignment  \\\n",
       "Sample                                                         \n",
       "BBL6672-PFG3-LFG2-R385-BUG   s__Bifidobacterium adolescentis   \n",
       "BBL6672-PFG3-LFG2-R385-BUG2                             None   \n",
       "\n",
       "                                                                Query taxonomy  \\\n",
       "Sample                                                                           \n",
       "BBL6672-PFG3-LFG2-R385-BUG   d__Bacteria;p__Actinobacteriota;c__Actinobacte...   \n",
       "BBL6672-PFG3-LFG2-R385-BUG2                                               None   \n",
       "\n",
       "                            Closest representative      ANI        AF  \\\n",
       "Sample                                                                  \n",
       "BBL6672-PFG3-LFG2-R385-BUG              G000010425  97.8854  0.857759   \n",
       "BBL6672-PFG3-LFG2-R385-BUG2                   None     None      None   \n",
       "\n",
       "                                                       Representative taxonomy  \n",
       "Sample                                                                          \n",
       "BBL6672-PFG3-LFG2-R385-BUG   d__Bacteria;p__Actinobacteriota;c__Actinobacte...  \n",
       "BBL6672-PFG3-LFG2-R385-BUG2                                               None  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_id='gg'\n",
    "new_df = []\n",
    "sample_IDs=['BBL6672-PFG3-LFG2-R385-BUG', 'BBL6672-PFG3-LFG2-R385-BUG2']\n",
    "cols = ['Query species assignment', 'Query taxonomy', 'Closest representative', 'ANI', 'AF', 'Representative taxonomy']\n",
    "for sample_ID in sample_IDs:\n",
    "    summ_df = pd.read_csv(f'/Users/liam/Downloads/testa/{sample_ID}.sp_classification.tsv', sep=\"\\t\", index_col=0)\n",
    "    if 'None' in summ_df.columns:\n",
    "        #print(111,pd.Series({k: 'None' for k in cols}, name=sample_ID), '\\n')\n",
    "        new_df.append(pd.Series({k: 'None' for k in cols}, name=sample_ID))\n",
    "    else:\n",
    "        #print(2222,summ_df.T[sample_ID])\n",
    "        new_df.append(summ_df.T[sample_ID])\n",
    "df = pd.DataFrame(new_df)\n",
    "df.index.name = 'Sample'\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "a    1\n",
       "b    2\n",
       "c    3\n",
       "Name: gg, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = {'a': 1, 'b': 2, 'c': 3}\n",
    "\n",
    "ser = pd.Series({'a': 1, 'b': 2, 'c': 3}, name='gg')\n",
    "\n",
    "ser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N50</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Contamination</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AssemblySummary1</th>\n",
       "      <td>1389366</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AssemblySummary2</th>\n",
       "      <td>1399366</td>\n",
       "      <td>90.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      N50 Completeness Contamination\n",
       "Sample                                              \n",
       "AssemblySummary1  1389366       100.00          0.00\n",
       "AssemblySummary2  1399366        90.00          0.00"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p='/Users/liam/Downloads/testa/'\n",
    "samples= ['AssemblySummary1','AssemblySummary2']\n",
    "df = pd.DataFrame(list())\n",
    "for sample_id in samples:\n",
    "    with open(p+sample_id+'.txt') as f:\n",
    "        for line in f:\n",
    "            line_split = line.split()\n",
    "            if len(line_split) == 0:\n",
    "                continue\n",
    "            if line_split[0] == '#':\n",
    "                continue\n",
    "            if line_split[0].lower() == 'contamination':\n",
    "                df.loc[sample_id, 'Contamination'] = line_split[-1]\n",
    "            elif line_split[0].lower() == 'completeness':\n",
    "                df.loc[sample_id, 'Completeness'] = line_split[-1]\n",
    "            elif line_split[0].upper() == 'N50':\n",
    "                if sample_id in df.index:\n",
    "                    if pd.isnull(df.loc[sample_id, 'N50']):\n",
    "                        df.loc[sample_id, 'N50'] = line_split[-1]\n",
    "                else:\n",
    "                    df.loc[sample_id, 'N50'] = line_split[-1]\n",
    "df.index.name = 'Sample'\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AssemblySummary1 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366'}})\n",
      "AssemblySummary1 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366'}})\n",
      "AssemblySummary1 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366'}})\n",
      "AssemblySummary1 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00'}})\n",
      "AssemblySummary1 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00', 'Contamination': '0.00'}})\n",
      "AssemblySummary2 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00', 'Contamination': '0.00'}, 'AssemblySummary2': {'N50': '1399366111'}})\n",
      "AssemblySummary2 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00', 'Contamination': '0.00'}, 'AssemblySummary2': {'N50': '1399366111'}})\n",
      "AssemblySummary2 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00', 'Contamination': '0.00'}, 'AssemblySummary2': {'N50': '1399366111'}})\n",
      "AssemblySummary2 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00', 'Contamination': '0.00'}, 'AssemblySummary2': {'N50': '1399366111', 'Completeness': '90.00'}})\n",
      "AssemblySummary2 defaultdict(<class 'dict'>, {'AssemblySummary1': {'N50': '1389366', 'Completeness': '100.00', 'Contamination': '0.00'}, 'AssemblySummary2': {'N50': '1399366111', 'Completeness': '90.00', 'Contamination': '0.00'}})\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N50</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Contamination</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AssemblySummary1</th>\n",
       "      <td>1389366</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AssemblySummary2</th>\n",
       "      <td>1399366111</td>\n",
       "      <td>90.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         N50 Completeness Contamination\n",
       "Sample                                                 \n",
       "AssemblySummary1     1389366       100.00          0.00\n",
       "AssemblySummary2  1399366111        90.00          0.00"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = defaultdict(dict)\n",
    "col_names = ['contamination', 'completeness', 'N50']\n",
    "p='/Users/liam/Downloads/testa/'\n",
    "samples= ['AssemblySummary1','AssemblySummary2']\n",
    "for sample_id in samples:\n",
    "    with open(p+sample_id+'.txt') as f:\n",
    "        for line in f:\n",
    "            bits = line.strip().split()\n",
    "            if len(bits) == 2 and bits[0] in col_names:\n",
    "                d[sample_id].setdefault(bits[0].capitalize(), bits[1])\n",
    "                print(sample_id, d)\n",
    "df = pd.DataFrame.from_dict(d, orient='index')\n",
    "df.index.name = 'Sample'\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contamination</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>N50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AssemblySummary1</th>\n",
       "      <td>1389366</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AssemblySummary2</th>\n",
       "      <td>1399366</td>\n",
       "      <td>90.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Contamination Completeness   N50\n",
       "Sample                                           \n",
       "AssemblySummary1       1389366       100.00  0.00\n",
       "AssemblySummary2       1399366        90.00  0.00"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p='/Users/liam/Downloads/testa/'\n",
    "samples= ['AssemblySummary1','AssemblySummary2']\n",
    "col_names = ['contamination', 'completeness', 'N50']\n",
    "df = pd.concat([\n",
    "        pd.read_csv(\n",
    "                p+sample_id+'.txt',\n",
    "                sep='\\t',\n",
    "                skip_blank_lines=True,\n",
    "                index_col=None,\n",
    "                names=['foo', sample_id])\n",
    "        .query(f'foo in {col_names}')\n",
    "        .reset_index(drop=True)\n",
    "        .drop([1,2])\n",
    "        .set_index('foo') \n",
    "        .T\n",
    "        for sample_id in samples\n",
    "        ])\n",
    "df.columns = map(str.capitalize, col_names)\n",
    "df.index.name = 'Sample'\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>a</th>\n",
       "      <th>N50</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AssemblySummary2</th>\n",
       "      <td>1399366</td>\n",
       "      <td>90.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "a                     N50 completeness contamination\n",
       "AssemblySummary2  1399366        90.00          0.00"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\n",
    "    p+sample_id+'.txt',\n",
    "    sep='\\t',\n",
    "    skip_blank_lines=True,\n",
    "    index_col=None,\n",
    "    names=['a', sample_id]) \\\n",
    "    .query(f'a in {col_names}') \\\n",
    "    .reset_index(drop=True) \\\n",
    "    .drop([1,2]) \\\n",
    "    .set_index('a') \\\n",
    "    .T\n",
    "#df.index = ['sa']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df = pd.concat([\n",
    "    pd.read_csv(assembly_summary, comment='#', sep=' ') for sample_id, assembly_summary in assembly_files.items()\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/liam/Downloads/r394_release_IDs.txt') as fin:\n",
    "    ids = []\n",
    "    for line in fin:\n",
    "        ids.append(line.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/liam/Downloads/r394_all_ids.txt') as fin, open('/Users/liam/Downloads/r394_fin.txt', 'w') as fout:\n",
    "    for line in fin:\n",
    "        if line.strip().split('_')[0] in ids:\n",
    "            fout.write(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(103,\n",
       " ['BBC7235-PFH0-LFH0-R396-FSS',\n",
       "  'BBC9835-PFH0-LFH0-R396-FSS',\n",
       "  'BBD1787-PFH0-LFH0-R396-FSS',\n",
       "  'BBF2314-PFH0-LFH0-R396-FSS'],\n",
       " ['BBC7235', 'BBC9835', 'BBD1787', 'BBF2314'])"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/Users/liam/Downloads/MBS-1342_attachments/R396.commands_PST.sh', 'r') as fin:\n",
    "    ids103=[]\n",
    "    bb_ids103=[]\n",
    "    for line in fin:\n",
    "        ids103.append(line.strip().split()[7])\n",
    "        bb_ids103.append(line.strip().split()[7].split('-')[0])\n",
    "len(ids103), ids103[:4], bb_ids103[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('/Users/liam/Downloads/ids_94.txt', 'r') as fin:\n",
    "    ids94=[]\n",
    "    for line in fin:\n",
    "        ids94.append(line.strip())\n",
    "len(ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(ids94).difference(set(bb_ids103))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BBJ7165',\n",
       " 'BBJ7283',\n",
       " 'BBK9725',\n",
       " 'BBK9727',\n",
       " 'BBK9732',\n",
       " 'BBK9737',\n",
       " 'BBK9756',\n",
       " 'BBK9775',\n",
       " 'BBK9816'}"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(bb_ids103).difference(set(ids94))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def a():\n",
    "    return 1,\n",
    "\n",
    "x, = a()\n",
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Color.RED: 1>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from enum import Enum\n",
    "class Color(Enum):\n",
    "    ['ff'] = 1\n",
    "    GREEN = 2\n",
    "    BLUE = 3\n",
    "\n",
    "Color(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "111 Season.SPRING\n",
      "Spring\n",
      "[<Season.SPRING: 1>, <Season.SUMMER: 2>, <Season.AUTUMN: 3>, <Season.WINTER: 4>]\n"
     ]
    }
   ],
   "source": [
    "from enum import Enum\n",
    "\n",
    "\n",
    "class Season(Enum):\n",
    "    SPRING = 1\n",
    "    SUMMER = 2\n",
    "    AUTUMN = 3\n",
    "    WINTER = 4\n",
    "\n",
    "\n",
    "seas = Season.SPRING\n",
    "print(111,seas)\n",
    "\n",
    "if seas == Season.SPRING:\n",
    "    print(\"Spring\")\n",
    "\n",
    "print(list(Season))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from enum import Enum\n",
    "class Color(Enum):\n",
    "    RED = 1\n",
    "    GREEN = 2\n",
    "    BLUE = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'REDa'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = Color['RED']\n",
    "a.name + 'a'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.10.2 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
